
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../assembly_evaluation/">
      
      
        <link rel="next" href="../bin_quality/">
      
      
      <link rel="icon" href="../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.5.14">
    
    
      
        <title>Binning - Metagenomics-Toolkit</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.10ba22f1.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="blue-grey" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#metabat" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      <div data-md-color-scheme="default" data-md-component="outdated" hidden>
        
          <aside class="md-banner md-banner--warning">
            <div class="md-banner__inner md-grid md-typeset">
              
  You're not viewing the latest version.
    <a href="../../../..">
	        <strong>Click here to go to latest.</strong>
    </a>

            </div>
            <script>var el=document.querySelector("[data-md-component=outdated]"),outdated=__md_get("__outdated",sessionStorage);!0===outdated&&el&&(el.hidden=!1)</script>
          </aside>
        
      </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="Metagenomics-Toolkit" class="md-header__button md-logo" aria-label="Metagenomics-Toolkit" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Metagenomics-Toolkit
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Binning
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="blue-grey" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="blue-grey" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/metagenomics/metagenomics-tk" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    Metagenomics-Tk
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../.." class="md-tabs__link">
        
  
    
  
  Introduction

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../../quickstart/" class="md-tabs__link">
        
  
    
  
  Quickstart

      </a>
    </li>
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../../overview/" class="md-tabs__link">
          
  
  Getting Started

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../../configuration/" class="md-tabs__link">
          
  
  Configuration

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../../modules/introduction/" class="md-tabs__link">
          
  
  Modules

        </a>
      </li>
    
  

      
        
  
  
    
  
  
    
    
      
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../introduction/" class="md-tabs__link">
          
  
  Tutorials

        </a>
      </li>
    
  

    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../../developer_guidelines/" class="md-tabs__link">
          
  
  Developer

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="Metagenomics-Toolkit" class="md-nav__button md-logo" aria-label="Metagenomics-Toolkit" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    Metagenomics-Toolkit
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/metagenomics/metagenomics-tk" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    Metagenomics-Tk
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../quickstart/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Quickstart
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Getting Started
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Getting Started
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../overview/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../full_pipeline/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Full Pipeline
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../aggregation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Aggregation
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Configuration
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Configuration
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../configuration/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Global Pipeline Configuration
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../pipeline_input/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Pipeline Input Configuration
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../database/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Database Configuration
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Modules
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Modules
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/introduction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/qualityControl/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Quality Control
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/annotation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Annotation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/assembly/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Assembly
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/cooccurrence/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Cooccurrence
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/dereplication/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Dereplication
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/fragment_recruitment/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Run Fragment Recruitment
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/magAttributes/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    MagAttributes
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/plasmids/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Plasmids
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/metabolomics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Metabolomics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/readMapping/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Read Mapping
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/export/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Exploratory Metagenome Browser (EMGB)
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
      
        
        
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6" checked>
        
          
          <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Tutorials
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_6">
            <span class="md-nav__icon md-icon"></span>
            Tutorials
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6_1" checked>
        
          
          <label class="md-nav__link" for="__nav_6_1" id="__nav_6_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Tutorial (Basics)
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_6_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_6_1">
            <span class="md-nav__icon md-icon"></span>
            Tutorial (Basics)
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../introduction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../quality_control/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Quality Control
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../assembly/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Assembly
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../assembly_evaluation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Assembly Evaluation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Binning
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Binning
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#metabat" class="md-nav__link">
    <span class="md-ellipsis">
      MetaBAT
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#assembly-evaluation-by-read-mapping" class="md-nav__link">
    <span class="md-ellipsis">
      Assembly Evaluation by Read Mapping
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#metabat-results" class="md-nav__link">
    <span class="md-ellipsis">
      MetaBAT Results
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../bin_quality/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Assessing Bin Quality
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../classification/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Classification
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../annotation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Annotation
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_7" >
        
          
          <label class="md-nav__link" for="__nav_7" id="__nav_7_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Developer
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_7">
            <span class="md-nav__icon md-icon"></span>
            Developer
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../developer_guidelines/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Guidelines
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../pipeline_specification/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Pipeline Specification
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../module_specification/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Module Specification
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#metabat" class="md-nav__link">
    <span class="md-ellipsis">
      MetaBAT
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#assembly-evaluation-by-read-mapping" class="md-nav__link">
    <span class="md-ellipsis">
      Assembly Evaluation by Read Mapping
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#metabat-results" class="md-nav__link">
    <span class="md-ellipsis">
      MetaBAT Results
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


  <h1>Binning</h1>

<p>We will now perform binning of our assembly.</p>
<p>Binning is a critical process in metagenomics that involves grouping DNA fragments, known as contigs,
into bins that likely originate from the same organism or genome.
This technique is essential for reconstructing genomes from mixed microbial communities,
where DNA from various organisms is intermingled,
and there is often no prior knowledge of the species present.</p>
<p>The binning process primarily utilizes two types of information: composition and coverage of the assembled contigs.</p>
<ul>
<li>
<p><strong>Composition:</strong> This refers to the analysis of k-mer frequency profiles within the contigs. K-mers are short DNA sequences of a fixed length (commonly tetranucleotides, or sequences of four nucleotides). These k-mer frequencies are generally conserved within a single genome, meaning that different regions of the same genome will share similar sequence composition.</p>
</li>
<li>
<p><strong>Coverage:</strong> This reflects the abundance of each contig in the assembly. Organisms that are more abundant in the environment contribute more DNA to the metagenomic sample,
resulting in their sequences being represented more frequently in the data.</p>
</li>
</ul>
<p>During binning, binning tools look for DNA fragments that, despite not being assembled together,
exhibit similar composition and occur at approximately equal abundances.
These similarities suggest that such contigs likely originate from the same genome,
enabling accurate grouping into bins.</p>
<p>By integrating both composition and coverage information, 
binning provides a powerful approach to distinguish between different organisms within a mixed sample. This method is particularly valuable as it allows researchers to reconstruct genomes without prior knowledge of the microbial community, making it an indispensable tool in metagenomic studies.</p>
<h2 id="metabat">MetaBAT<a class="headerlink" href="#metabat" title="Permanent link">&para;</a></h2>
<p>MetaBAT is an automated metagenome binning software
which integrates empirical probabilistic distances of genome abundance
and tetranucleotide frequency. See the <a href="https://bitbucket.org/berkeleylab/metabat">MetaBAT home page</a> for more information.</p>
<p>We will now perform a binning of our assembly with MetaBAT using the Metagenomics-Toolkit.
The following snippet represents the Toolkit configuration for the binning module:</p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">Binning Configuration File Snippet 1</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="w">  </span><span class="nt">binning</span><span class="p">:</span>
<span class="w">    </span><span class="nt">bwa2</span><span class="p">:</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span><span class="w"> </span>
<span class="w">        </span><span class="nt">bwa2</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">        </span><span class="c1"># samtools flags are used to filter resulting bam file</span>
<span class="w">        </span><span class="nt">samtoolsView</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">-F</span><span class="nv"> </span><span class="s">3584</span><span class="nv"> </span><span class="s">&quot;</span><span class="w"> </span>
<span class="w">    </span><span class="nt">contigsCoverage</span><span class="p">:</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">--min-covered-fraction</span><span class="nv"> </span><span class="s">0</span><span class="nv"> </span><span class="s">--min-read-percent-identity</span><span class="nv"> </span><span class="s">100</span><span class="nv"> </span><span class="s">--min-read-aligned-percent</span><span class="nv"> </span><span class="s">100</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">    </span><span class="nt">genomeCoverage</span><span class="p">:</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">--min-covered-fraction</span><span class="nv"> </span><span class="s">0</span><span class="nv"> </span><span class="s">--min-read-percent-identity</span><span class="nv"> </span><span class="s">100</span><span class="nv"> </span><span class="s">--min-read-aligned-percent</span><span class="nv"> </span><span class="s">100</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">    </span><span class="c1"># Primary binning tool</span>
<span class="w">    </span><span class="nt">metabat</span><span class="p">:</span>
<span class="w">      </span><span class="c1"># Set --seed positive numbers to reproduce the result exactly. Otherwise, random seed will be set each time.</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">--seed</span><span class="nv"> </span><span class="s">234234</span><span class="nv">  </span><span class="s">&quot;</span>
</code></pre></div></td></tr></table></div>
<div class="admonition question">
<p class="admonition-title">Task 1</p>
<p>We can run it directly with:
<div class="highlight"><pre><span></span><code><span class="nb">cd</span><span class="w"> </span>~/mgcourse/

<span class="nv">NXF_VER</span><span class="o">=</span><span class="m">24</span>.10.4<span class="w"> </span>nextflow<span class="w"> </span>run<span class="w"> </span>metagenomics/metagenomics-tk<span class="w"> </span><span class="se">\</span>
<span class="w">      </span>-profile<span class="w"> </span>standard<span class="w"> </span><span class="se">\</span>
<span class="w">      </span>-params-file<span class="w"> </span>https://raw.githubusercontent.com/metagenomics/metagenomics-tk/refs/heads/master/default/tutorials/tutorial1/fullpipeline_binning.yml<span class="w"> </span><span class="se">\</span>
<span class="w">      </span>-ansi-log<span class="w"> </span><span class="nb">false</span><span class="w"> </span><span class="se">\</span>
<span class="w">      </span>-entry<span class="w"> </span>wFullPipeline<span class="w"> </span><span class="se">\</span>
<span class="w">      </span>-resume<span class="w"> </span><span class="se">\</span>
<span class="w">      </span>--input.paired.path<span class="w"> </span>https://raw.githubusercontent.com/metagenomics/metagenomics-tk/refs/heads/master/test_data/tutorials/tutorial1/reads.tsv<span class="w"> </span><span class="se">\</span>
<span class="w">      </span>--logDir<span class="w"> </span>logs_binning<span class="w"> </span>
</code></pre></div></p>
</div>
<p>For reference, your complete parameter file looks like this:</p>
<details class="parameter-file">
<summary>Parameter-file</summary>
<div class="highlight"><pre><span></span><code>tempdir:<span class="w"> </span><span class="s2">&quot;tmp&quot;</span>
summary:<span class="w"> </span><span class="nb">false</span>
s3SignIn:<span class="w"> </span><span class="nb">false</span>
input:
<span class="w">  </span>paired:
<span class="w">    </span>path:<span class="w"> </span><span class="s2">&quot;test_data/tutorials/tutorial1/reads.tsv&quot;</span>
<span class="w">    </span>watch:<span class="w"> </span><span class="nb">false</span>
output:<span class="w"> </span>output
logDir:<span class="w"> </span>log
runid:<span class="w"> </span><span class="m">1</span>
databases:<span class="w"> </span><span class="s2">&quot;/vol/scratch/databases&quot;</span>
publishDirMode:<span class="w"> </span><span class="s2">&quot;symlink&quot;</span>
logLevel:<span class="w"> </span><span class="m">1</span>
scratch:<span class="w"> </span><span class="nb">false</span>
steps:
<span class="w">  </span>qc:
<span class="w">    </span>fastp:
<span class="w">       </span><span class="c1"># For PE data, the adapter sequence auto-detection is disabled by default since the adapters can be trimmed by overlap analysis. However, you can specify --detect_adapter_for_pe to enable it.</span>
<span class="w">       </span><span class="c1"># For PE data, fastp will run a little slower if you specify the sequence adapters or enable adapter auto-detection, but usually result in a slightly cleaner output, since the overlap analysis may fail due to sequencing errors or adapter dimers.</span>
<span class="w">       </span><span class="c1"># -q, --qualified_quality_phred       the quality value that a base is qualified. Default 15 means phred quality &gt;=Q15 is qualified.</span>
<span class="w">       </span><span class="c1"># --cut_front move a sliding window from front (5&#39;) to tail, drop the bases in the window if its mean quality is below cut_mean_quality, stop otherwise.</span>
<span class="w">       </span><span class="c1"># --length_required  reads shorter than length_required will be discarded, default is 15. (int [=15])</span>
<span class="w">       </span><span class="c1"># PE data, the front/tail trimming settings are given with -f, --trim_front1 and -t, --trim_tail1</span>
<span class="w">       </span>additionalParams:
<span class="w">         </span>fastp:<span class="w"> </span><span class="s2">&quot; --detect_adapter_for_pe -q 20 --cut_front --trim_front1 3 --cut_tail --trim_tail1 3 --cut_mean_quality 10 --length_required 50 &quot;</span>
<span class="w">         </span>reportOnly:<span class="w"> </span><span class="nb">false</span>
<span class="w">       </span>timeLimit:<span class="w"> </span><span class="s2">&quot;AUTO&quot;</span>
<span class="w">    </span>nonpareil:
<span class="w">      </span>additionalParams:<span class="w"> </span><span class="s2">&quot; -v 10 -r 1234 &quot;</span>
<span class="w">    </span>filterHuman:
<span class="w">      </span>additionalParams:<span class="w"> </span><span class="s2">&quot;  &quot;</span>
<span class="w">      </span>database:
<span class="w">        </span>download:
<span class="w">          </span>source:<span class="w"> </span>https://openstack.cebitec.uni-bielefeld.de:8080/databases/human_filter.db.20231218v2.gz
<span class="w">          </span>md5sum:<span class="w"> </span>cc92c0f926656565b1156d66a0db5a3c
<span class="w">  </span>assembly:
<span class="w">    </span>megahit:
<span class="w">      </span><span class="c1"># --mem-flag 0 to use minimum memory, --mem-flag 1 (default) moderate memory and --mem-flag 2 all memory.</span>
<span class="w">      </span><span class="c1"># meta-sensitive: &#39;--min-count 1 --k-list 21,29,39,49,...,129,141&#39; </span>
<span class="w">      </span><span class="c1"># meta-large:  &#39;--k-min  27  --k-max 127 --k-step 10&#39; (large &amp; complex metagenomes, like soil)</span>
<span class="w">      </span>additionalParams:<span class="w"> </span><span class="s2">&quot; --min-contig-len 500 --presets meta-sensitive &quot;</span>
<span class="w">      </span>fastg:<span class="w"> </span><span class="nb">true</span>
<span class="w">      </span>resources:
<span class="w">        </span>RAM:
<span class="w">          </span>mode:<span class="w"> </span><span class="s1">&#39;PREDICT&#39;</span>
<span class="w">          </span>predictMinLabel:<span class="w"> </span><span class="s1">&#39;highmemLarge&#39;</span>
<span class="w">  </span>binning:
<span class="w">    </span>bwa2:
<span class="w">      </span>additionalParams:<span class="w"> </span>
<span class="w">        </span>bwa2:<span class="w"> </span><span class="s2">&quot; &quot;</span>
<span class="w">        </span><span class="c1"># samtools flags are used to filter resulting bam file</span>
<span class="w">        </span>samtoolsView:<span class="w"> </span><span class="s2">&quot; -F 3584 &quot;</span><span class="w"> </span>
<span class="w">    </span>contigsCoverage:
<span class="w">      </span>additionalParams:<span class="w"> </span><span class="s2">&quot; --min-covered-fraction 0 --min-read-percent-identity 100 --min-read-aligned-percent 100 &quot;</span>
<span class="w">    </span>genomeCoverage:
<span class="w">      </span>additionalParams:<span class="w"> </span><span class="s2">&quot; --min-covered-fraction 0 --min-read-percent-identity 100 --min-read-aligned-percent 100 &quot;</span>
<span class="w">    </span><span class="c1"># Primary binning tool</span>
<span class="w">    </span>metabat:
<span class="w">      </span><span class="c1"># Set --seed positive numbers to reproduce the result exactly. Otherwise, random seed will be set each time.</span>
<span class="w">      </span>additionalParams:<span class="w"> </span><span class="s2">&quot; --seed 234234  &quot;</span>
resources:
<span class="w">  </span>highmemLarge:
<span class="w">    </span>cpus:<span class="w"> </span><span class="m">28</span>
<span class="w">    </span>memory:<span class="w"> </span><span class="m">60</span>
<span class="w">  </span>highmemMedium:
<span class="w">    </span>cpus:<span class="w"> </span><span class="m">14</span>
<span class="w">    </span>memory:<span class="w"> </span><span class="m">30</span>
<span class="w">  </span>large:
<span class="w">    </span>cpus:<span class="w"> </span><span class="m">28</span>
<span class="w">    </span>memory:<span class="w"> </span><span class="m">58</span>
<span class="w">  </span>medium:
<span class="w">    </span>cpus:<span class="w"> </span><span class="m">14</span>
<span class="w">    </span>memory:<span class="w"> </span><span class="m">29</span>
<span class="w">  </span>small:
<span class="w">    </span>cpus:<span class="w"> </span><span class="m">7</span>
<span class="w">    </span>memory:<span class="w"> </span><span class="m">14</span>
<span class="w">  </span>tiny:
<span class="w">    </span>cpus:<span class="w"> </span><span class="m">1</span>
<span class="w">    </span>memory:<span class="w"> </span><span class="m">1</span>
</code></pre></div>
</details>
<h2 id="assembly-evaluation-by-read-mapping">Assembly Evaluation by Read Mapping<a class="headerlink" href="#assembly-evaluation-by-read-mapping" title="Permanent link">&para;</a></h2>
<p>Before we take a closer look at the binning results, 
let us first check the mapping of the reads to the assembly (alignment) that was computed as part of the binning process and saved in the file with the <code>.bam</code> prefix.
The number of reads mapped to the contigs of the assembly is called the contig abundance and is also used as input to a binning tool to separate contigs.
From this alignment we get the contig <strong>coverage</strong> (read depth) which refers to the number of times each nucleotide position in a contig is covered by sequencing reads on average.
It quantifies how many reads align to each base pair in the assembled sequence.</p>
<div class="admonition question">
<p class="admonition-title">Task 2</p>
<p>First we have to index the sorted BAM file:
<div class="highlight"><pre><span></span><code><span class="nb">cd</span><span class="w"> </span>~/mgcourse/
</code></pre></div>
<div class="highlight"><pre><span></span><code>samtools<span class="w"> </span>index<span class="w"> </span>output/data/1/binning/*/contigMapping/data.bam
</code></pre></div></p>
</div>
<p>Usually a bam file has to be sorted before it can be indexed. The sorting has already been done for you by the Toolkit.</p>
<div class="admonition question">
<p class="admonition-title">Task 3</p>
<p>To look at the BAM file you can use the following command (You can quit less with <code>q</code>):
<div class="highlight"><pre><span></span><code><span class="nb">cd</span><span class="w"> </span>~/mgcourse/
</code></pre></div>
<div class="highlight"><pre><span></span><code>samtools<span class="w"> </span>view<span class="w"> </span>output/data/1/binning/*/contigMapping/data.bam<span class="w"> </span><span class="p">|</span><span class="w"> </span>less
</code></pre></div></p>
</div>
<p>IGV, or Integrated Genome Viewer, is an open-source bioinformatics tool designed for the interactive visualization and exploration of genomic data.
It primarily serves as a platform to view alignments of sequencing reads against a reference genome,
aiding in the identification of genetic variants such as SNPs (Single Nucleotide Polymorphisms) and indels (insertions and deletions).
Beyond read alignment, IGV supports various data types, including gene annotations and expression levels, 
offering a comprehensive view that enhances understanding of the genomic context.
In our case, we are only interested in the alignment.</p>
<div class="admonition question">
<p class="admonition-title">Task 4</p>
<p>Now copy/link everything you need for igv in a separate folder:</p>
<p><div class="highlight"><pre><span></span><code><span class="nb">cd</span><span class="w"> </span>~/mgcourse/
</code></pre></div>
<div class="highlight"><pre><span></span><code>mkdir<span class="w"> </span>igv_data
<span class="nb">cd</span><span class="w"> </span>igv_data
cp<span class="w"> </span>../output/data/1/assembly/*/megahit/data_contigs.fa.gz<span class="w"> </span>.
gunzip<span class="w"> </span>data_contigs.fa.gz
ln<span class="w"> </span>-s<span class="w"> </span>../output/data/1/binning/*/contigMapping/data.bam
ln<span class="w"> </span>-s<span class="w"> </span>../output/data/1/binning/*/contigMapping/data.bam.bai
</code></pre></div></p>
</div>
<div class="admonition question">
<p class="admonition-title">Task 5</p>
<p>We will use the IGV genome browser to look at the mappings.
<div class="highlight"><pre><span></span><code>igv
</code></pre></div></p>
<p>Now let's look at the mapped reads:</p>
<ol>
<li>Load the contig sequences into IGV. Use the menu <code>Genomes-&gt;Load Genome from File...</code> and search for <code>home|ubuntu/mgcourse/igv_data/data_contigs.fa</code></li>
<li>Load the BAM file into IGV. Use menu <code>File-&gt;Load from File...</code> and search for <code>home|ubuntu/mgcourse/igv_data/data.bam</code></li>
</ol>
</div>
<div class="admonition question">
<p class="admonition-title">Task 6</p>
<p>Look for errors in the mappings - are all those error sequencing errors?</p>
<details class="solution">
<summary>Solution</summary>
<p>Some errors are due to merging multiple strains into one contig. This can be clearly seen 
when there are a large number of errors at one position with different possibilities for that base.</p>
</details>
</div>
<h2 id="metabat-results">MetaBAT Results<a class="headerlink" href="#metabat-results" title="Permanent link">&para;</a></h2>
<p>Let's now inspect the bins created by MetaBAT:</p>
<div class="admonition question">
<p class="admonition-title">Task 7</p>
<p>How many bins did metabat generate? Locate the metabat results and the fasta files for each bin in the output folder.</p>
<details class="solution">
<summary>Solution</summary>
<p><div class="highlight"><pre><span></span><code><span class="nb">cd</span><span class="w"> </span>~/mgcourse/
</code></pre></div>
<div class="highlight"><pre><span></span><code>ls<span class="w"> </span>-l<span class="w"> </span>output/data/1/binning/*/metabat/
</code></pre></div>
<div class="highlight"><pre><span></span><code>data_bin.1.fa
data_bin.10.fa
data_bin.2.fa
data_bin.3.fa
data_bin.4.fa
data_bin.5.fa
data_bin.6.fa
data_bin.7.fa
data_bin.8.fa
data_bin.9.fa
data_bin_contig_mapping.tsv
data_bins_stats.tsv
data_contigs_depth.tsv
data_notBinned.fa
</code></pre></div>
MetaBAT generated 10 bins.</p>
</details>
</div>
<p>You might have noticed that there is also a <code>data_notBinned.fa</code> file that represents contigs that could not be binned. We will investigate these contigs in the
next section of this tutorial.
The next question you might want to ask is whether you can trust these bins and which organism they represent according to a taxonomy.
Before that, let's have a look at what other information the Metagenomics-Toolkit provides as part of the binning output.</p>
<div class="admonition question">
<p class="admonition-title">Task 8</p>
<p>There is a file containing some statistics on the generated bins <code>data_bins_stats.tsv</code>. 
Find out, which bin has the highest coverage and which one has the highest N50.</p>
<details class="solution">
<summary>Solution</summary>
<p>You can have a look on some bin statistics with:
<div class="highlight"><pre><span></span><code>cd ~/mgcourse/
</code></pre></div>
<div class="highlight"><pre><span></span><code>less<span class="w"> </span>output/data/1/binning/0.5.0/metabat/data_bins_stats.tsv
</code></pre></div>
<div class="highlight"><pre><span></span><code>SAMPLE<span class="w">  </span>BIN_ID<span class="w">  </span>format<span class="w">  </span><span class="nb">type</span><span class="w">    </span>num_seqs<span class="w">        </span>sum_len<span class="w"> </span>min_len<span class="w"> </span>avg_len<span class="w"> </span>max_len<span class="w"> </span>Q1<span class="w">      </span>Q2<span class="w">      </span>Q3<span class="w">      </span>sum_gap<span class="w"> </span>N50<span class="w">     </span>Q20<span class="o">(</span>%<span class="o">)</span><span class="w">  </span>Q30<span class="o">(</span>%<span class="o">)</span><span class="w">  </span>GC<span class="o">(</span>%<span class="o">)</span><span class="w">   </span>COVERAGE
data<span class="w">    </span>data_bin.1.fa<span class="w">   </span>FASTA<span class="w">   </span>DNA<span class="w">     </span><span class="m">150</span><span class="w">     </span><span class="m">507900</span><span class="w">  </span><span class="m">2504</span><span class="w">    </span><span class="m">3386</span>.0<span class="w">  </span><span class="m">7616</span><span class="w">    </span><span class="m">2757</span>.0<span class="w">  </span><span class="m">3127</span>.5<span class="w">  </span><span class="m">3704</span>.0<span class="w">  </span><span class="m">0</span><span class="w">       </span><span class="m">3321</span><span class="w">    </span><span class="m">0</span>.00<span class="w">    </span><span class="m">0</span>.00<span class="w">    </span><span class="m">33</span>.29<span class="w">   </span><span class="m">6</span>.15009
data<span class="w">    </span>data_bin.10.fa<span class="w">  </span>FASTA<span class="w">   </span>DNA<span class="w">     </span><span class="m">177</span><span class="w">     </span><span class="m">641723</span><span class="w">  </span><span class="m">2504</span><span class="w">    </span><span class="m">3625</span>.6<span class="w">  </span><span class="m">8372</span><span class="w">    </span><span class="m">2855</span>.0<span class="w">  </span><span class="m">3213</span>.0<span class="w">  </span><span class="m">4036</span>.0<span class="w">  </span><span class="m">0</span><span class="w">       </span><span class="m">3542</span><span class="w">    </span><span class="m">0</span>.00<span class="w">    </span><span class="m">0</span>.00<span class="w">    </span><span class="m">49</span>.49<span class="w">   </span><span class="m">7</span>.7348
data<span class="w">    </span>data_bin.2.fa<span class="w">   </span>FASTA<span class="w">   </span>DNA<span class="w">     </span><span class="m">224</span><span class="w">     </span><span class="m">717735</span><span class="w">  </span><span class="m">2500</span><span class="w">    </span><span class="m">3204</span>.2<span class="w">  </span><span class="m">6592</span><span class="w">    </span><span class="m">2692</span>.0<span class="w">  </span><span class="m">2958</span>.5<span class="w">  </span><span class="m">3442</span>.0<span class="w">  </span><span class="m">0</span><span class="w">       </span><span class="m">3109</span><span class="w">    </span><span class="m">0</span>.00<span class="w">    </span><span class="m">0</span>.00<span class="w">    </span><span class="m">50</span>.72<span class="w">   </span><span class="m">5</span>.05946
data<span class="w">    </span>data_bin.3.fa<span class="w">   </span>FASTA<span class="w">   </span>DNA<span class="w">     </span><span class="m">150</span><span class="w">     </span><span class="m">520975</span><span class="w">  </span><span class="m">2500</span><span class="w">    </span><span class="m">3473</span>.2<span class="w">  </span><span class="m">9084</span><span class="w">    </span><span class="m">2739</span>.0<span class="w">  </span><span class="m">3127</span>.0<span class="w">  </span><span class="m">3909</span>.0<span class="w">  </span><span class="m">0</span><span class="w">       </span><span class="m">3328</span><span class="w">    </span><span class="m">0</span>.00<span class="w">    </span><span class="m">0</span>.00<span class="w">    </span><span class="m">43</span>.93<span class="w">   </span><span class="m">7</span>.0744
data<span class="w">    </span>data_bin.4.fa<span class="w">   </span>FASTA<span class="w">   </span>DNA<span class="w">     </span><span class="m">264</span><span class="w">     </span><span class="m">1340562</span><span class="w"> </span><span class="m">2505</span><span class="w">    </span><span class="m">5077</span>.9<span class="w">  </span><span class="m">17160</span><span class="w">   </span><span class="m">3213</span>.0<span class="w">  </span><span class="m">4328</span>.5<span class="w">  </span><span class="m">6132</span>.5<span class="w">  </span><span class="m">0</span><span class="w">       </span><span class="m">5540</span><span class="w">    </span><span class="m">0</span>.00<span class="w">    </span><span class="m">0</span>.00<span class="w">    </span><span class="m">32</span>.36<span class="w">   </span><span class="m">10</span>.8599
data<span class="w">    </span>data_bin.5.fa<span class="w">   </span>FASTA<span class="w">   </span>DNA<span class="w">     </span><span class="m">88</span><span class="w">      </span><span class="m">280467</span><span class="w">  </span><span class="m">2509</span><span class="w">    </span><span class="m">3187</span>.1<span class="w">  </span><span class="m">5229</span><span class="w">    </span><span class="m">2746</span>.5<span class="w">  </span><span class="m">2998</span>.5<span class="w">  </span><span class="m">3429</span>.5<span class="w">  </span><span class="m">0</span><span class="w">       </span><span class="m">3093</span><span class="w">    </span><span class="m">0</span>.00<span class="w">    </span><span class="m">0</span>.00<span class="w">    </span><span class="m">39</span>.02<span class="w">   </span><span class="m">5</span>.12779
data<span class="w">    </span>data_bin.6.fa<span class="w">   </span>FASTA<span class="w">   </span>DNA<span class="w">     </span><span class="m">139</span><span class="w">     </span><span class="m">1149027</span><span class="w"> </span><span class="m">2523</span><span class="w">    </span><span class="m">8266</span>.4<span class="w">  </span><span class="m">35033</span><span class="w">   </span><span class="m">4367</span>.0<span class="w">  </span><span class="m">6419</span>.0<span class="w">  </span><span class="m">10607</span>.0<span class="w"> </span><span class="m">0</span><span class="w">       </span><span class="m">10241</span><span class="w">   </span><span class="m">0</span>.00<span class="w">    </span><span class="m">0</span>.00<span class="w">    </span><span class="m">40</span>.52<span class="w">   </span><span class="m">14</span>.8691
data<span class="w">    </span>data_bin.7.fa<span class="w">   </span>FASTA<span class="w">   </span>DNA<span class="w">     </span><span class="m">280</span><span class="w">     </span><span class="m">944933</span><span class="w">  </span><span class="m">2505</span><span class="w">    </span><span class="m">3374</span>.8<span class="w">  </span><span class="m">6899</span><span class="w">    </span><span class="m">2786</span>.5<span class="w">  </span><span class="m">3165</span>.0<span class="w">  </span><span class="m">3690</span>.5<span class="w">  </span><span class="m">0</span><span class="w">       </span><span class="m">3346</span><span class="w">    </span><span class="m">0</span>.00<span class="w">    </span><span class="m">0</span>.00<span class="w">    </span><span class="m">57</span>.96<span class="w">   </span><span class="m">6</span>.31446
data<span class="w">    </span>data_bin.8.fa<span class="w">   </span>FASTA<span class="w">   </span>DNA<span class="w">     </span><span class="m">66</span><span class="w">      </span><span class="m">682946</span><span class="w">  </span><span class="m">2837</span><span class="w">    </span><span class="m">10347</span>.7<span class="w"> </span><span class="m">31045</span><span class="w">   </span><span class="m">6083</span>.0<span class="w">  </span><span class="m">8277</span>.5<span class="w">  </span><span class="m">13778</span>.0<span class="w"> </span><span class="m">0</span><span class="w">       </span><span class="m">13397</span><span class="w">   </span><span class="m">0</span>.00<span class="w">    </span><span class="m">0</span>.00<span class="w">    </span><span class="m">25</span>.03<span class="w">   </span><span class="m">15</span>.588
data<span class="w">    </span>data_bin.9.fa<span class="w">   </span>FASTA<span class="w">   </span>DNA<span class="w">     </span><span class="m">374</span><span class="w">     </span><span class="m">1625795</span><span class="w"> </span><span class="m">2501</span><span class="w">    </span><span class="m">4347</span>.0<span class="w">  </span><span class="m">11747</span><span class="w">   </span><span class="m">3019</span>.0<span class="w">  </span><span class="m">3725</span>.5<span class="w">  </span><span class="m">5239</span>.0<span class="w">  </span><span class="m">0</span><span class="w">       </span><span class="m">4532</span><span class="w">    </span><span class="m">0</span>.00<span class="w">    </span><span class="m">0</span>.00<span class="w">    </span><span class="m">27</span>.53<span class="w">   </span><span class="m">9</span>.6491
</code></pre></div>
In this result, bin 8 has the highest coverage (15.588) and highest N50 (13397). Also note, that some of the bins highly differ in their GC content.</p>
</details>
</div>
<div class="admonition question">
<p class="admonition-title">Task 9</p>
<p>Let's say we want to take a closer look at these bins. Switch back to EMGB and try to find the bin with the largest contig, this one:
<div class="highlight"><pre><span></span><code>SAMPLE  BIN_ID  format  type    num_seqs        sum_len min_len avg_len max_len Q1      Q2      Q3      sum_gap N50     Q20(%)  Q30(%)  GC(%)   COVERAGE
data    data_bin.4.fa   FASTA   DNA     139     1149027 2523    8266.4  35033   4367.0  6419.0  10607.0 0       10241   0.00    0.00    40.52   14.9085
</code></pre></div>
How many genes are found on this contig and how long is the longest one on it?</p>
<details class="solution">
<summary>Solution</summary>
<p>One way to get to the desired contig is to click on the "Contigs" tab in the upper middle part of EMGB. It is located between "Genes" and "GO Stats".
Here we activate the filter "Show only contigs longer than x" and set the filter to 30000. Two contigs remain, the longer one at the top is the one we are looking for. 
The table shows us the first solution, right next to the contig name and length it says that there are 41 genes on this contig.        </p>
<p>If you look to the right of our contig, you will see the "Actions" tab with two buttons. A colorful one representing small contigs and a Download ".fna" button.
Click on the first one. A popup will open, showing this contig with all predicted genes and their positions on this contig. If you look at the top of the popup, 
you will find the solution to your second problem. There are some basic statistics for this contig:
<div class="highlight"><pre><span></span><code>Overall length (bp)     Number of genes     Longest gene (bp)   Dataset     Current gene
35033   41  2756    data    -   
</code></pre></div>
The longest gene is 2756 bp in size.</p>
</details>
</div>
<div class="admonition question">
<p class="admonition-title">Task 10</p>
<p>We now know what the longest contig looks like, but what about its bin? Go back to EMGB again and try to find the name of the bin our contig belongs to.
How many contigs are in that bin?</p>
<details class="solution">
<summary>Solution</summary>
<p>An easy way to find our contig is to use filters again. Simply search for the longest contig again, as we did in the task above. This time, we don't click on the "full contigs" 
button, but on the little filter icon right next to the contig name, you might see a hovering text "create a filter for this contig".
After clicking it, we see a new filter at the top left "Contig ID is exactly X", which means that only information that this contig is part of will be shown.
We now click on the "MAGs" (Metagenome-assembled genomes) tab, three tabs to the right of the "Contigs" tab we are currently in. Since only information about this contig is shown, 
thanks to the filter set, and the contig can only be part of one bin/MAG, there is only one left. The one we are looking for, our contig, is part of "bin 7".
Lets do the same trick again. Click on the filter icon next to the bin name. There should now be another active filter at the top left, "Bins is exactly X".
Go back to the Contigs tab. Only or first contig in the list? Sure, because there is still our first filter that lets us see only information about this contig.
Go to the top filter section and remove the first filter "Contig ID is exactly X" by clicking the little X in the top right corner of its box. With the filter removed, 
there is only "Bins is exactly X" left, which means we see all information about this bin. If not, is the "Show only contigs longer than X" box still checked? If so, uncheck it. 
With this done, the information should update, we see "Showing 139 contigs that contain predicted genes matching the filters", which is the answer to our question. </p>
</details>
</div>
<hr />
<p>➡️ <a href="../bin_quality/"><strong>Continue to: Assessing Bin Quality</strong></a></p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../../..", "features": ["navigation.tabs", "navigation.tabs.sticky", "content.code.copy"], "search": "../../../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"provider": "mike"}}</script>
    
    
      <script src="../../../assets/javascripts/bundle.bd41221c.min.js"></script>
      
    
  </body>
</html>