
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../../../modules/export/">
      
      
        <link rel="next" href="../quality_control/">
      
      
      <link rel="icon" href="../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.5.14">
    
    
      
        <title>Introduction - Metagenomics-Toolkit</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.10ba22f1.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="blue-grey" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#tutorial-scope-and-requirements" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      <div data-md-color-scheme="default" data-md-component="outdated" hidden>
        
          <aside class="md-banner md-banner--warning">
            <div class="md-banner__inner md-grid md-typeset">
              
  You're not viewing the latest version.
    <a href="../../../..">
	        <strong>Click here to go to latest.</strong>
    </a>

            </div>
            <script>var el=document.querySelector("[data-md-component=outdated]"),outdated=__md_get("__outdated",sessionStorage);!0===outdated&&el&&(el.hidden=!1)</script>
          </aside>
        
      </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="Metagenomics-Toolkit" class="md-header__button md-logo" aria-label="Metagenomics-Toolkit" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Metagenomics-Toolkit
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Introduction
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="blue-grey" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="blue-grey" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/metagenomics/metagenomics-tk" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    Metagenomics-Tk
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../.." class="md-tabs__link">
        
  
    
  
  Introduction

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../../quickstart/" class="md-tabs__link">
        
  
    
  
  Quickstart

      </a>
    </li>
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../../overview/" class="md-tabs__link">
          
  
  Getting Started

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../../configuration/" class="md-tabs__link">
          
  
  Configuration

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../../modules/introduction/" class="md-tabs__link">
          
  
  Modules

        </a>
      </li>
    
  

      
        
  
  
    
  
  
    
    
      
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="./" class="md-tabs__link">
          
  
  Tutorials

        </a>
      </li>
    
  

    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../../developer_guidelines/" class="md-tabs__link">
          
  
  Developer

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="Metagenomics-Toolkit" class="md-nav__button md-logo" aria-label="Metagenomics-Toolkit" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    Metagenomics-Toolkit
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/metagenomics/metagenomics-tk" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    Metagenomics-Tk
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../quickstart/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Quickstart
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Getting Started
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Getting Started
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../overview/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../full_pipeline/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Full Pipeline
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../aggregation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Aggregation
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Configuration
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Configuration
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../configuration/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Global Pipeline Configuration
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../pipeline_input/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Pipeline Input Configuration
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../database/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Database Configuration
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Modules
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Modules
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/introduction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/qualityControl/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Quality Control
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/annotation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Annotation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/assembly/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Assembly
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/cooccurrence/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Cooccurrence
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/dereplication/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Dereplication
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/fragment_recruitment/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Run Fragment Recruitment
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/magAttributes/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    MagAttributes
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/plasmids/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Plasmids
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/metabolomics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Metabolomics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/readMapping/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Read Mapping
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/export/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Exploratory Metagenome Browser (EMGB)
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
      
        
        
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6" checked>
        
          
          <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Tutorials
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_6">
            <span class="md-nav__icon md-icon"></span>
            Tutorials
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6_1" checked>
        
          
          <label class="md-nav__link" for="__nav_6_1" id="__nav_6_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Tutorial (Basics)
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_6_1_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_6_1">
            <span class="md-nav__icon md-icon"></span>
            Tutorial (Basics)
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Introduction
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Introduction
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#tutorial-scope-and-requirements" class="md-nav__link">
    <span class="md-ellipsis">
      Tutorial Scope and Requirements
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Tutorial Scope and Requirements">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#requirements" class="md-nav__link">
    <span class="md-ellipsis">
      Requirements
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#preparations" class="md-nav__link">
    <span class="md-ellipsis">
      Preparations
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Preparations">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#link-to-your-volume" class="md-nav__link">
    <span class="md-ellipsis">
      Link to your volume
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#metagenomics-toolkit-introduction" class="md-nav__link">
    <span class="md-ellipsis">
      Metagenomics-Toolkit Introduction
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Metagenomics-Toolkit Introduction">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#execution" class="md-nav__link">
    <span class="md-ellipsis">
      Execution
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#configuration" class="md-nav__link">
    <span class="md-ellipsis">
      Configuration
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Configuration">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#part-1-global-workflow-parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Part 1: Global Workflow Parameters
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Part 1: Global Workflow Parameters">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#input-field" class="md-nav__link">
    <span class="md-ellipsis">
      Input Field
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#part-2-toolkit-analyses-steps" class="md-nav__link">
    <span class="md-ellipsis">
      Part 2: Toolkit Analyses Steps
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#part-3-computational-resources" class="md-nav__link">
    <span class="md-ellipsis">
      Part 3: Computational Resources
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#configuration-file-vs-commandline-parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Configuration File vs. Commandline Parameters
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#output" class="md-nav__link">
    <span class="md-ellipsis">
      Output
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../quality_control/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Quality Control
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../assembly/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Assembly
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../assembly_evaluation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Assembly Evaluation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../binning/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Binning
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../bin_quality/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Assessing Bin Quality
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../classification/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Classification
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../annotation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Annotation
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_7" >
        
          
          <label class="md-nav__link" for="__nav_7" id="__nav_7_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Developer
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_7">
            <span class="md-nav__icon md-icon"></span>
            Developer
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../developer_guidelines/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Guidelines
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../pipeline_specification/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Pipeline Specification
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../module_specification/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Module Specification
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#tutorial-scope-and-requirements" class="md-nav__link">
    <span class="md-ellipsis">
      Tutorial Scope and Requirements
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Tutorial Scope and Requirements">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#requirements" class="md-nav__link">
    <span class="md-ellipsis">
      Requirements
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#preparations" class="md-nav__link">
    <span class="md-ellipsis">
      Preparations
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Preparations">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#link-to-your-volume" class="md-nav__link">
    <span class="md-ellipsis">
      Link to your volume
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#metagenomics-toolkit-introduction" class="md-nav__link">
    <span class="md-ellipsis">
      Metagenomics-Toolkit Introduction
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Metagenomics-Toolkit Introduction">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#execution" class="md-nav__link">
    <span class="md-ellipsis">
      Execution
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#configuration" class="md-nav__link">
    <span class="md-ellipsis">
      Configuration
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Configuration">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#part-1-global-workflow-parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Part 1: Global Workflow Parameters
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Part 1: Global Workflow Parameters">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#input-field" class="md-nav__link">
    <span class="md-ellipsis">
      Input Field
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#part-2-toolkit-analyses-steps" class="md-nav__link">
    <span class="md-ellipsis">
      Part 2: Toolkit Analyses Steps
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#part-3-computational-resources" class="md-nav__link">
    <span class="md-ellipsis">
      Part 3: Computational Resources
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#configuration-file-vs-commandline-parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Configuration File vs. Commandline Parameters
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#output" class="md-nav__link">
    <span class="md-ellipsis">
      Output
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


  <h1>Introduction</h1>

<p>This workshop demonstrates typical steps of a short-read metagenomic analysis using the Metagenomics-Toolkit and is divided into eight parts.
In this part you will learn how to configure and run the Toolkit and what the output of a Toolkit run looks like.</p>
<h2 id="tutorial-scope-and-requirements">Tutorial Scope and Requirements<a class="headerlink" href="#tutorial-scope-and-requirements" title="Permanent link">&para;</a></h2>
<p>The Metagenomics-Toolkit allows you to run either the full pipeline of assembly, binning and many other downstream analysis tasks or the individual analyses separately. 
In this tutorial you will only use the <em>full pipeline</em> mode. The <em>full pipeline</em> mode itself is structured into two parts. The first part runs the Toolkit on each
sample separately (<em>per-sample</em>), and the second part runs a combined downstream analysis on the output of the <em>per-sample</em> part, called <em>aggregation</em>. 
In this tutorial, you will only run the <em>per-sample</em> part. While there are several optimizations for running the Toolkit on a cloud-based setup, 
during this workshop you will run the Toolkit on a single machine.</p>
<h3 id="requirements">Requirements<a class="headerlink" href="#requirements" title="Permanent link">&para;</a></h3>
<div class="admonition warning">
<p class="admonition-title">Course Participants</p>
<p>In case you are a course participant, it is very likely that a machine has been prepared for you and you can ignore this section.
If in doubt, ask the course organizers. </p>
</div>
<ul>
<li>
<p>Basic Linux command line usage</p>
</li>
<li>
<p>This tutorial has been tested on a machine with 28 CPUs and 64 GBs of RAM with Ubuntu installed on it.</p>
</li>
<li>
<p>Docker: Install Docker by following the official Docker installation <a href="https://docs.docker.com/engine/install/ubuntu/">instructions</a>.</p>
</li>
<li>
<p>Java: In order to run Nextflow, you need to install Java on your machine, which can be achieved via <code>sudo apt install default-jre</code>.</p>
</li>
<li>
<p>Nextflow should be installed. Please check the official Nextflow <a href="https://www.nextflow.io/docs/latest/install.html#install-nextflow">instructions</a>.</p>
</li>
</ul>
<h3 id="preparations">Preparations<a class="headerlink" href="#preparations" title="Permanent link">&para;</a></h3>
<p>For this tutorial, we need some preparations on our machine.
<strong>Please note that the following instructions can be customized for your specific machine.</strong> </p>
<h4 id="link-to-your-volume">Link to your volume<a class="headerlink" href="#link-to-your-volume" title="Permanent link">&para;</a></h4>
<p>We are using the location <code>~/mgcourse/</code> as the directory, where all the analysis is stored. We assume that there is a larger storage volume mounted at <code>/vol/mgcourse</code>. 
If your preferred volume is at another location, change <code>/vol/mgcourse</code> to a different path accordingly. If you don't have another volume with more storage available
you can also simply leave out the following command and create the directory with <code>mkdir ~/mgcourse</code>.</p>
<div class="admonition warning">
<p class="admonition-title">Course Participants</p>
<p>In case you are a course participant, a volume has been prepared for you, so please run the following commands. </p>
</div>
<div class="admonition question">
<p class="admonition-title">Task 1</p>
<p>We will link the volume to <code>~/mgcourse</code>, so we can use that link for the rest of this tutorial:
<div class="highlight"><pre><span></span><code>ln<span class="w"> </span>-s<span class="w"> </span>/vol/mgcourse/<span class="w"> </span>~/mgcourse
</code></pre></div></p>
<p>Create a database directory</p>
<div class="highlight"><pre><span></span><code>mkdir<span class="w"> </span>-p<span class="w"> </span>~/mgcourse/databases
</code></pre></div>
</div>
<h2 id="metagenomics-toolkit-introduction">Metagenomics-Toolkit Introduction<a class="headerlink" href="#metagenomics-toolkit-introduction" title="Permanent link">&para;</a></h2>
<h3 id="execution">Execution<a class="headerlink" href="#execution" title="Permanent link">&para;</a></h3>
<p>The Toolkit is based on Nextflow and you can execute the Toolkit based on the following command line schema:</p>
<div class="highlight"><pre><span></span><code><span class="nv">NXF_VER</span><span class="o">=</span>NEXTFLOW_VERSION<span class="w"> </span>nextflow<span class="w"> </span>run<span class="w"> </span>metagenomics/metagenomics-tk<span class="w"> </span>NEXTFLOW_OPTIONS<span class="w"> </span>TOOLKIT_OPTIONS<span class="w"> </span>
</code></pre></div>
<ul>
<li>
<p><code>NEXTFLOW_VERSION</code> is the Nextflow version that is supported by Nextflow. Every code snippet in this tutorial has a hard coded version number. 
  If you ever choose the wrong version, then the Toolkit will print out the versions that are supported.</p>
</li>
<li>
<p><code>NEXTFLOW_OPTIONS</code> are options that are implemented by Nextflow:</p>
<ul>
<li>
<p><code>-profile</code> determines the technology which is used to execute the Toolkit. Here we support <strong>standard</strong> for running the workflow on a single machine and
             <strong>slurm</strong> for running the Toolkit on a cluster which uses <a href="https://slurm.schedmd.com/documentation.html">SLURM</a> to distribute jobs. </p>
</li>
<li>
<p><code>-params-file</code> points to a configuration file that tells the Toolkit which analyses to run and which resources it should use. An example configuration file will be explained in the next section.</p>
</li>
<li>
<p><code>-resume</code> In some cases, you may want to resume the workflow execution, such as when you add an analysis.
              Resuming the workflow forces Nextflow to reuse the results of the previous analyses that the new analysis depends on, rather than starting from scratch. </p>
</li>
<li>
<p><code>-ansi-log</code> accepts a boolean (default: <strong>true</strong>) that tells Nextflow on <strong>true</strong> to print every update as a new line on the terminal. If <strong>false</strong> then Nextflow
              prints a line for every process and updates the specific line on an update. We recommend to setting <strong>-ansi-log</strong> to <strong>false</strong> because it is not possible to
              print all possible processes on a terminal at once when running the Toolkit.</p>
</li>
<li>
<p><code>-entry</code> specifies which entrypoint Nextflow should use to run the workflow. For running the <em>full pipeline</em>, that you will use in this workshop, you use the <strong>wFullPipeline</strong> entrypoint. 
           If you ever want to run separate modules, you can check on the modules specific page (e.g. <a href="../../../modules/assembly/#assembly">assembly</a>).</p>
</li>
</ul>
</li>
<li>
<p><code>TOOLKIT_OPTIONS</code> are options that are provided by the Toolkit. All Toolkit options are either in a configuration file or can be provided on the command line which will be explained in the following section. </p>
</li>
</ul>
<div class="admonition question">
<p class="admonition-title">Task 2</p>
<p>Open the Metagenomics-Toolkit wiki on a second browser tab by a click on this
<a href="https://metagenomics.github.io/metagenomics-tk/latest/" target="_blank">link</a>.
Imagine you need to run the quality-control part separately. Can you tell the name of the <strong>entrypoint</strong>? 
Use the wiki page you have opened on another tab to answer the question.</p>
<details class="solution">
<summary>Solution</summary>
<p>If you go to the <a href="../../../modules/qualityControl/#quality-control">quality control</a> part, then you will find
the <strong>wShortReadQualityControl</strong> entrypoint for short reads and the <strong>wOntQualityControl</strong> entrypoint for long
reads.</p>
</details>
</div>
<h3 id="configuration">Configuration<a class="headerlink" href="#configuration" title="Permanent link">&para;</a></h3>
<p>The Toolkit uses a YAML configuration file that specifies global parameters, the analyses that will be executed and the computational resources that can be used. </p>
<p>The configuration file is divided into three parts:</p>
<h4 id="part-1-global-workflow-parameters">Part 1: Global Workflow Parameters<a class="headerlink" href="#part-1-global-workflow-parameters" title="Permanent link">&para;</a></h4>
<p>The following snippet shows parameters that affect the whole execution of the 
workflow. All parameters are explained in a dedicated Toolkit wiki <a href="../../../configuration/">section</a>. </p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">Example Configuration File Snippet 1</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nt">summary</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">false</span>
<span class="nt">s3SignIn</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">false</span><span class="w"> </span>
<span class="nt">input</span><span class="p">:</span>
<span class="w">  </span><span class="nt">paired</span><span class="p">:</span>
<span class="w">    </span><span class="nt">path</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;test_data/tutorials/tutorial1/reads.tsv&quot;</span>
<span class="w">    </span><span class="nt">watch</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">false</span>
<span class="nt">output</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">output</span>
<span class="nt">logDir</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">log</span>
<span class="nt">runid</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1</span>
<span class="nt">databases</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;/vol/scratch/databases&quot;</span>
<span class="nt">publishDirMode</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;symlink&quot;</span>
<span class="nt">logLevel</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1</span>
<span class="nt">scratch</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">false</span><span class="w"> </span>
</code></pre></div></td></tr></table></div>
<div class="admonition info">
<p class="admonition-title">Computational Resources</p>
<p>Please note that computational resources are also global parameters and will be handled in the third part of this configuration section. </p>
</div>
<h5 id="input-field">Input Field<a class="headerlink" href="#input-field" title="Permanent link">&para;</a></h5>
<p>The input field (line 3, snippet 1) specifies the type of input data to process (Nanopore, Illumina, data hosted on SRA or a mirror)
and you can find a dedicated wiki section <a href="../../../pipeline_input/">here</a>. Regardless of which input type
is used, the user must provide a file containing a list of datasets to be processed.
The list can be a list of remote or local files and in the case of SRA, a list of SRA run IDs.</p>
<p>Since you will work with short read data in this tutorial, your input file looks like this:</p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">Input File</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div><pre><span></span><code>SAMPLE<span class="w">  </span>READS1<span class="w">  </span>READS2
data<span class="w">    </span>https://s3.bi.denbi.de/cmg/mgcourses/mg2025/reads/read1.fq.gz<span class="w">   </span>https://s3.bi.denbi.de/cmg/mgcourses/mg2025/reads/read2.fq.gz
</code></pre></div></td></tr></table></div>
<p>The first column (SAMPLE) specifies the name of the dataset. The second (READS1) and third (READS2) column specify the files
containing the forward and reverse reads.</p>
<h4 id="part-2-toolkit-analyses-steps">Part 2: Toolkit Analyses Steps<a class="headerlink" href="#part-2-toolkit-analyses-steps" title="Permanent link">&para;</a></h4>
<p>Analyses, or sometimes called modules, that the Toolkit executes are placed directly under the <strong>steps</strong> attribute in the configuration file.
In the example below, the modules <strong>qc</strong> and <strong>assembly</strong> are placed directly under the <strong>steps</strong> attribute. Any tools or methods
that are used as part of the module can be considered as a property of the module. For example <strong>MEGAHIT</strong> is executed as part of the assembly module.
The level below the tool names is for configuring the tools and methods. Each analysis is listed on the <a href="../../../modules/introduction/">modules page</a>. </p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">Example Configuration File Snippet 2</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nt">steps</span><span class="p">:</span>
<span class="w">  </span><span class="nt">qc</span><span class="p">:</span>
<span class="w">    </span><span class="nt">fastp</span><span class="p">:</span>
<span class="w">       </span><span class="c1"># For PE data, the adapter sequence auto-detection is disabled by default since the adapters can be trimmed by overlap analysis. However, you can specify --detect_adapter_for_pe to enable it.</span>
<span class="w">       </span><span class="c1"># For PE data, fastp will run a little slower if you specify the sequence adapters or enable adapter auto-detection, but usually result in a slightly cleaner output, since the overlap analysis may fail due to sequencing errors or adapter dimers.</span>
<span class="w">       </span><span class="c1"># -q, --qualified_quality_phred       the quality value that a base is qualified. Default 15 means phred quality &gt;=Q15 is qualified.</span>
<span class="w">       </span><span class="c1"># --cut_front move a sliding window from front (5&#39;) to tail, drop the bases in the window if its mean quality is below cut_mean_quality, stop otherwise.</span>
<span class="w">       </span><span class="c1"># --length_required  reads shorter than length_required will be discarded, default is 15. (int [=15])</span>
<span class="w">       </span><span class="c1"># PE data, the front/tail trimming settings are given with -f, --trim_front1 and -t, --trim_tail1</span>
<span class="w">       </span><span class="nt">additionalParams</span><span class="p">:</span>
<span class="w">         </span><span class="nt">fastp</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">--detect_adapter_for_pe</span><span class="nv"> </span><span class="s">-q</span><span class="nv"> </span><span class="s">20</span><span class="nv"> </span><span class="s">--cut_front</span><span class="nv"> </span><span class="s">--trim_front1</span><span class="nv"> </span><span class="s">3</span><span class="nv"> </span><span class="s">--cut_tail</span><span class="nv"> </span><span class="s">--trim_tail1</span><span class="nv"> </span><span class="s">3</span><span class="nv"> </span><span class="s">--cut_mean_quality</span><span class="nv"> </span><span class="s">10</span><span class="nv"> </span><span class="s">--length_required</span><span class="nv"> </span><span class="s">50</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">         </span><span class="nt">reportOnly</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">false</span>
<span class="w">       </span><span class="nt">timeLimit</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;AUTO&quot;</span>
<span class="w">    </span><span class="nt">nonpareil</span><span class="p">:</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">-v</span><span class="nv"> </span><span class="s">10</span><span class="nv"> </span><span class="s">-r</span><span class="nv"> </span><span class="s">1234</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">    </span><span class="nt">filterHuman</span><span class="p">:</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv">  </span><span class="s">&quot;</span>
<span class="w">      </span><span class="nt">database</span><span class="p">:</span>
<span class="w">        </span><span class="nt">download</span><span class="p">:</span>
<span class="w">          </span><span class="nt">source</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">https://openstack.cebitec.uni-bielefeld.de:8080/databases/human_filter.db.20231218v2.gz</span>
<span class="w">          </span><span class="nt">md5sum</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">cc92c0f926656565b1156d66a0db5a3c</span>
<span class="w">  </span><span class="nt">assembly</span><span class="p">:</span>
<span class="w">    </span><span class="nt">megahit</span><span class="p">:</span>
<span class="w">      </span><span class="c1"># --mem-flag 0 to use minimum memory, --mem-flag 1 (default) moderate memory and --mem-flag 2 all memory.</span>
<span class="w">      </span><span class="c1"># meta-sensitive: &#39;--min-count 1 --k-list 21,29,39,49,...,129,141&#39; </span>
<span class="w">      </span><span class="c1"># meta-large:  &#39;--k-min  27  --k-max 127 --k-step 10&#39; (large &amp; complex metagenomes, like soil)</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">--min-contig-len</span><span class="nv"> </span><span class="s">500</span><span class="nv"> </span><span class="s">--presets</span><span class="nv"> </span><span class="s">meta-sensitive</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">      </span><span class="nt">fastg</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">true</span>
<span class="w">      </span><span class="nt">resources</span><span class="p">:</span>
<span class="w">        </span><span class="nt">RAM</span><span class="p">:</span>
<span class="w">          </span><span class="nt">mode</span><span class="p">:</span><span class="w"> </span><span class="s">&#39;PREDICT&#39;</span>
<span class="w">          </span><span class="nt">predictMinLabel</span><span class="p">:</span><span class="w"> </span><span class="s">&#39;highmemLarge&#39;</span>
</code></pre></div></td></tr></table></div>
<h4 id="part-3-computational-resources">Part 3: Computational Resources<a class="headerlink" href="#part-3-computational-resources" title="Permanent link">&para;</a></h4>
<p>The third part of a Toolkit configuration file is the <strong>resources</strong> attribute.
The <strong>resources</strong> attribute lists computational resource configurations, where each configuration has a label and consists of the number of CPUs and amount of RAM assigned to it.
Predefined labels are listed in the following example snippet. These labels are assigned to the processes that run the workflow specific tools.
You can read more  about resource parameters <a href="../../../configuration/#configuration-of-computational-resources-used-for-pipeline-runs">here</a>.</p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">Example Configuration File Snippet 3</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nt">resources</span><span class="p">:</span>
<span class="w">  </span><span class="nt">highmemLarge</span><span class="p">:</span>
<span class="w">    </span><span class="nt">cpus</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">28</span>
<span class="w">    </span><span class="nt">memory</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">60</span>
<span class="w">  </span><span class="nt">highmemMedium</span><span class="p">:</span>
<span class="w">    </span><span class="nt">cpus</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">14</span>
<span class="w">    </span><span class="nt">memory</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">30</span>
<span class="w">  </span><span class="nt">large</span><span class="p">:</span>
<span class="w">    </span><span class="nt">cpus</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">28</span>
<span class="w">    </span><span class="nt">memory</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">58</span>
<span class="w">  </span><span class="nt">medium</span><span class="p">:</span>
<span class="w">    </span><span class="nt">cpus</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">14</span>
<span class="w">    </span><span class="nt">memory</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">29</span>
<span class="w">  </span><span class="nt">small</span><span class="p">:</span>
<span class="w">    </span><span class="nt">cpus</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">7</span>
<span class="w">    </span><span class="nt">memory</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">14</span>
<span class="w">  </span><span class="nt">tiny</span><span class="p">:</span>
<span class="w">    </span><span class="nt">cpus</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1</span>
<span class="w">    </span><span class="nt">memory</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1</span>
</code></pre></div></td></tr></table></div>
<div class="admonition question">
<p class="admonition-title">Task 3</p>
<p>One of the first checks before running the Toolkit is to adjust the resource labels to the resources of your machine.
You can run <code>nproc</code> to get the number of CPUs and <code>free -h --mega</code> to get the amount of RAM (row name: Mem, column name: total) available on your machine. 
Is there enough RAM on your machine to run the Toolkit? </p>
<details class="solution">
<summary>Solution</summary>
<p>If you are using a machine as described in the "Requirements" section, then yes, there is enough RAM available for the workflow (more than 60 GBs). </p>
</details>
</div>
<h4 id="configuration-file-vs-commandline-parameters">Configuration File vs. Commandline Parameters<a class="headerlink" href="#configuration-file-vs-commandline-parameters" title="Permanent link">&para;</a></h4>
<p>All parameters defined in the YAML configuration file can also be supplied as command-line arguments. To do this, prefix each parameter with a double dash (--). 
If a parameter is nested within the hierarchy of the YAML file, represent it as a command-line argument by connecting each level of the hierarchy using a dot (.).</p>
<p>For example, consider the CPU count of the <em>highmemLarge</em> resource label in the previous snippet.
The corresponding command-line argument would be <code>--resources.highmemLarge.cpus</code>.</p>
<div class="admonition question">
<p class="admonition-title">Task 4</p>
<p>Let`s say you want to specify a path to a different input TSV file (see <strong>Example Configuration File Snippet 1</strong>) that contains a different set of input datasets. 
How would you specify the parameter on the command line?</p>
<details class="solution">
<summary>Solution</summary>
<p><code>--input.paired.path</code></p>
</details>
</div>
<p>Command line arguments supersede the configuration file. This is a quick, temporary way to change variables without touching files. </p>
<h3 id="output">Output<a class="headerlink" href="#output" title="Permanent link">&para;</a></h3>
<p>The Toolkit output fulfills the following schema:</p>
<div class="highlight"><pre><span></span><code>SAMPLE_NAME/RUN_ID/MODULE/MODULE_VERSION/TOOL
</code></pre></div>
<ul>
<li>
<p><strong>RUN_ID:</strong> The run ID will be part of the output path and allows to distinguish between different pipeline configurations that were used for the same dataset.</p>
</li>
<li>
<p><strong>MODULE</strong> is the analysis that is executed by the Toolkit (e.g. qc, assembly, etc.). </p>
</li>
<li>
<p><strong>MODULE_VERSION</strong> is the version number of the module.</p>
</li>
<li>
<p><strong>TOOL</strong> is the tool or method that is executed by the Toolkit.</p>
</li>
</ul>
<p>Below you can see an example output structure.
Every output folder includes four log files:</p>
<ul>
<li>
<p><strong>.command.err:</strong> Contains the standard error.</p>
</li>
<li>
<p><strong>.command.out:</strong> Contains the standard output. </p>
</li>
<li>
<p><strong>.command.log:</strong> Contains the combined standard error and standard output.</p>
</li>
<li>
<p><strong>.command.sh:</strong> Contains the command that was executed. </p>
</li>
</ul>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">Example Output Directory</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span></pre></div></td><td class="code"><div><pre><span></span><code>output/
└──<span class="w"> </span>sample
<span class="w">    </span>└──<span class="w"> </span><span class="m">1</span>
<span class="w">        </span>└──<span class="w"> </span>qc
<span class="w">            </span>└──<span class="w"> </span><span class="m">0</span>.3.0
<span class="w">                </span>├──<span class="w"> </span>fastp
<span class="w">                </span>│<span class="w">   </span>├──<span class="w"> </span>.command.err
<span class="w">                </span>│<span class="w">   </span>├──<span class="w"> </span>.command.log
<span class="w">                </span>│<span class="w">   </span>├──<span class="w"> </span>.command.out
<span class="w">                </span>│<span class="w">   </span>├──<span class="w"> </span>.command.sh
<span class="w">                </span>│<span class="w">   </span>├──<span class="w"> </span>sample_fastp.json
<span class="w">                </span>│<span class="w">   </span>├──<span class="w"> </span>sample_fastp_summary_after.tsv
<span class="w">                </span>│<span class="w">   </span>├──<span class="w"> </span>sample_fastp_summary_before.tsv
<span class="w">                </span>│<span class="w">   </span>├──<span class="w"> </span>sample_interleaved.qc.fq.gz
<span class="w">                </span>│<span class="w">   </span>├──<span class="w"> </span>sample_report.html
<span class="w">                </span>│<span class="w">   </span>├──<span class="w"> </span>sample_unpaired.qc.fq.gz
<span class="w">                </span>│<span class="w">   </span>└──<span class="w"> </span>sample_unpaired_summary.tsv
<span class="w">                </span>├──<span class="w"> </span>filterHuman
<span class="w">                </span>│<span class="w">   </span>├──<span class="w"> </span>.command.err
<span class="w">                </span>│<span class="w">   </span>├──<span class="w"> </span>.command.log
<span class="w">                </span>│<span class="w">   </span>├──<span class="w"> </span>.command.out
<span class="w">                </span>│<span class="w">   </span>├──<span class="w"> </span>.command.sh
<span class="w">                </span>│<span class="w">   </span>├──<span class="w"> </span>sample_interleaved.filtered.fq.gz
<span class="w">                </span>│<span class="w">   </span>├──<span class="w"> </span>sample_interleaved.removed.fq.gz
<span class="w">                </span>│<span class="w">   </span>├──<span class="w"> </span>sample_interleaved_summary_after.tsv
<span class="w">                </span>│<span class="w">   </span>├──<span class="w"> </span>sample_interleaved_summary_before.tsv
<span class="w">                </span>│<span class="w">   </span>├──<span class="w"> </span>sample_unpaired.filtered.fq.gz
<span class="w">                </span>│<span class="w">   </span>├──<span class="w"> </span>sample_unpaired_summary_after.tsv
<span class="w">                </span>│<span class="w">   </span>└──<span class="w"> </span>sample_unpaired_summary_before.tsv
<span class="w">                </span>├──<span class="w"> </span>kmc
<span class="w">                </span>│<span class="w">   </span>├──<span class="w"> </span>.command.err
<span class="w">                </span>│<span class="w">   </span>├──<span class="w"> </span>.command.log
<span class="w">                </span>│<span class="w">   </span>├──<span class="w"> </span>.command.out
<span class="w">                </span>│<span class="w">   </span>├──<span class="w"> </span>.command.sh
<span class="w">                </span>│<span class="w">   </span>├──<span class="w"> </span>sample.13.histo.tsv
<span class="w">                </span>│<span class="w">   </span>├──<span class="w"> </span>sample.13.kmc.json
<span class="w">                </span>│<span class="w">   </span>├──<span class="w"> </span>sample.21.histo.tsv
<span class="w">                </span>│<span class="w">   </span>├──<span class="w"> </span>sample.21.kmc.json
<span class="w">                </span>│<span class="w">   </span>├──<span class="w"> </span>sample.71.histo.tsv
<span class="w">                </span>│<span class="w">   </span>└──<span class="w"> </span>sample.71.kmc.json
<span class="w">                </span>└──<span class="w"> </span>nonpareil
<span class="w">                    </span>├──<span class="w"> </span>.command.err
<span class="w">                    </span>├──<span class="w"> </span>.command.log
<span class="w">                    </span>├──<span class="w"> </span>.command.out
<span class="w">                    </span>├──<span class="w"> </span>.command.sh
<span class="w">                    </span>├──<span class="w"> </span>sample.npa
<span class="w">                    </span>├──<span class="w"> </span>sample.npc
<span class="w">                    </span>├──<span class="w"> </span>sample.npl
<span class="w">                    </span>├──<span class="w"> </span>sample_nonpareil_curves.pdf
<span class="w">                    </span>└──<span class="w"> </span>sample_nonpareil_index.tsv
</code></pre></div></td></tr></table></div>
<hr />
<p>➡️ <a href="../quality_control/"><strong>Continue to: Quality Control</strong></a></p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../../..", "features": ["navigation.tabs", "navigation.tabs.sticky", "content.code.copy"], "search": "../../../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"provider": "mike"}}</script>
    
    
      <script src="../../../assets/javascripts/bundle.bd41221c.min.js"></script>
      
    
  </body>
</html>