# Meta-Omics-Toolkit: Output
## Quality Control
The output is a gzipped fastq file (short read: `SAMPLE_interleaved.qc.fq.gz`, long read: `SAMPLE_qc.fq.gz`) containing
trimmed and quality filtered reads.

## Assembly and Binning
The output is a gzipped fasta file containing contigs.

## Annotation
### MMseqs2
Calculated significant matches of a nucleotide/protein query which was compared against a user provided set of databases.

### MMseqs2 - Taxonomy
By identifying homologous through searches against a provided MMseqs2 taxonomy-database, MMseqs2 can compute the lowest
common ancestor. This lowest common ancestor is a robust taxonomic label for unknown sequences. These labels are presented
in form of an `*.taxonomy.tsv` file, a `*.krakenStyleTaxonomy.out` formatted in accordance to the
[KRAKEN](https://ccb.jhu.edu/software/kraken/) tool outputs and an interactive
[KRONA](https://github.com/marbl/Krona/wiki) plot in form of a html website `*.krona.html`.

### Prokka
Prokka computes `*.err`, `*.faa`, `*.ffn`, `*.fna`, `*.fsa`, `*.gbk`, `*.gff`, `*.sqn`, `*.tbl`, `*.tbl` for every bin.
`*.gbk` and `*.sqn` are skipped per default, since tbl2asn runs for quite a while! If you need those files generated by
prokka, include: `--tbl2asn` in the prokka parameters to enable it. Details of all files can be read on the Prokka page.
In addition, it also computes a summary tsv file which adheres to the magAttributes specification.

### KEGGFromBlast
Result `*.tsv` file filled with KEGG information (like modules, KO's, ...) which could be linked to the input hits.

### Resistance Gene Identifier (rgi)
The `*rgi.tsv` files contain the found CARD genes.

## Mag Attributes
### GTDBTk
All GTDB files include the GTDB specific columns in addition to a `SAMPLE` column (`SAMPLE_gtdbtk.bac120.summary.tsv`,
`SAMPLE_gtdbtk.ar122.summary.tsv`). In addition, this module produces a file `SAMPLE_gtdbtk_CHUNK.tsv` that combines
both files and adds a `BIN_ID` column that adheres to the magAttributes specification.

### Checkm
The Checkm output adheres to the magAttributes specification and adds a `BIN_ID` and `SAMPLE` column to the output file.

## Plasmids
### SCAPP
SCAPP detects plasmid sequences out of the samples assembly graph. It reports sequences as gzipped fasta files
(`*_plasmids.fasta.gz`). A basic statistic (`*_plasmids_stats.tsv`) per plasmid and a summary statistic
(`*_plasmids_summary_stats.tsv`) over all plasmids is also generated. Coverm coverage metrics are generated for all
plasmids. Gene coverage values are generated as part of the annotation module output.

### PlasClass
PlasClass is able to identify plasmids by using a statistical model that was build using kmer frequencies.
It reports gzipped fata files and their probabilities (`*_plasclass.tsv`).

### ViralVerifyPlasmid
ViralVerfiy is applying a Naive Bayes classifier (`*_viralverifyplasmid.tsv`).

### PLSDB
PLSDB includes a curated set of plasmid sequences that were extracted from databases like refseq. The metadata of
found sequences are reported in *.tsv and the metadata of the filtered sequences in `*_kmerThreshold_X.tsv`.
