
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../concept/">
      
      
        <link rel="next" href="../per_sample_and_aggregation/">
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.5.14">
    
    
      
        <title>Full Pipeline - Metagenomics-Toolkit</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.10ba22f1.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="blue-grey" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#requirements" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      <div data-md-color-scheme="default" data-md-component="outdated" hidden>
        
          <aside class="md-banner md-banner--warning">
            <div class="md-banner__inner md-grid md-typeset">
              
  You're not viewing the latest version.
    <a href="../..">
	        <strong>Click here to go to latest.</strong>
    </a>

            </div>
            <script>var el=document.querySelector("[data-md-component=outdated]"),outdated=__md_get("__outdated",sessionStorage);!0===outdated&&el&&(el.hidden=!1)</script>
          </aside>
        
      </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Metagenomics-Toolkit" class="md-header__button md-logo" aria-label="Metagenomics-Toolkit" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Metagenomics-Toolkit
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Full Pipeline
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="blue-grey" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="blue-grey" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    <li class="md-tabs__item">
      <a href=".." class="md-tabs__link">
        
  
    
  
  Introduction

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../quickstart/" class="md-tabs__link">
        
  
    
  
  Quickstart

      </a>
    </li>
  

      
        
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../concept/" class="md-tabs__link">
          
  
  Getting Started

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../configuration/" class="md-tabs__link">
          
  
  Configuration

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../modules/introduction/" class="md-tabs__link">
          
  
  Modules

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../developer_guidelines/" class="md-tabs__link">
          
  
  Developer

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Metagenomics-Toolkit" class="md-nav__button md-logo" aria-label="Metagenomics-Toolkit" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    Metagenomics-Toolkit
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../quickstart/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Quickstart
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
      
        
        
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" checked>
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Getting Started
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Getting Started
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../concept/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Concept
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Full Pipeline
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Full Pipeline
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#requirements" class="md-nav__link">
    <span class="md-ellipsis">
      Requirements
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#preparation" class="md-nav__link">
    <span class="md-ellipsis">
      Preparation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#run-the-toolkit" class="md-nav__link">
    <span class="md-ellipsis">
      Run the Toolkit
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Run the Toolkit">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#input" class="md-nav__link">
    <span class="md-ellipsis">
      Input
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#output" class="md-nav__link">
    <span class="md-ellipsis">
      Output
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#further-reading" class="md-nav__link">
    <span class="md-ellipsis">
      Further Reading
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../per_sample_and_aggregation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Per-Sample and Aggregation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../emgb/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    EMGB
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Configuration
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Configuration
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../configuration/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Global Pipeline Configuration
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../pipeline_input/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Pipeline Input Configuration
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../database/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Database Configuration
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Modules
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Modules
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../modules/introduction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../modules/qualityControl/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Quality Control
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../modules/annotation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Annotation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../modules/assembly/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Assembly
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../modules/cooccurrence/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Cooccurrence
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../modules/dereplication/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Dereplication
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../modules/fragment_recruitment/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Run Fragment Recruitment
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../modules/magAttributes/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    MagAttributes
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../modules/plasmids/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Plasmids
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../modules/metabolomics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Metabolomics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../modules/readMapping/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Read Mapping
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../modules/export/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Export
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6" >
        
          
          <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Developer
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6">
            <span class="md-nav__icon md-icon"></span>
            Developer
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../developer_guidelines/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Guidelines
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../pipeline_specification/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Pipeline Specification
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../module_specification/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Module Specification
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#requirements" class="md-nav__link">
    <span class="md-ellipsis">
      Requirements
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#preparation" class="md-nav__link">
    <span class="md-ellipsis">
      Preparation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#run-the-toolkit" class="md-nav__link">
    <span class="md-ellipsis">
      Run the Toolkit
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Run the Toolkit">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#input" class="md-nav__link">
    <span class="md-ellipsis">
      Input
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#output" class="md-nav__link">
    <span class="md-ellipsis">
      Output
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#further-reading" class="md-nav__link">
    <span class="md-ellipsis">
      Further Reading
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  <h1>Full Pipeline</h1>

<p>The full pipeline mode allows you to run the per-sample part and the aggregation part in one execution (see <a href="../">schematic overview</a>).
In contrast to the quickstart section, this chapter refers to the execution of the toolkit on a cluster system.</p>
<h2 id="requirements">Requirements<a class="headerlink" href="#requirements" title="Permanent link">&para;</a></h2>
<ul>
<li>
<p>SLURM: The Toolkit was mainly developed for cloud-based clusters using SLURM as a resource orchestrator.</p>
</li>
<li>
<p>Docker: Docker must be available on all worker nodes.</p>
</li>
<li>
<p>Java: In order to run Nextflow you have install Java.</p>
</li>
<li>
<p>Nextflow version: Please check in the nextflow.config file the supported Nextflow versions.</p>
</li>
<li>
<p>Resources: scratch</p>
</li>
</ul>
<h2 id="preparation">Preparation<a class="headerlink" href="#preparation" title="Permanent link">&para;</a></h2>
<p>First checkout the Github repository in a directory which is shared by all worker nodes:</p>
<div class="highlight"><pre><span></span><code>git<span class="w"> </span>clone<span class="w"> </span>git@github.com:metagenomics/metagenomics-tk.git
<span class="nb">cd</span><span class="w"> </span>metagenomics-tk
</code></pre></div>
<h2 id="run-the-toolkit">Run the Toolkit<a class="headerlink" href="#run-the-toolkit" title="Permanent link">&para;</a></h2>
<p>In the following case the Nextflow binary is placed on a share working directory:</p>
<div class="highlight"><pre><span></span><code>./nextflow<span class="w"> </span>run<span class="w"> </span>main.nf<span class="w"> </span>-work-dir<span class="w"> </span>/vol/spool/work<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-profile<span class="w"> </span>slurm<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-entry<span class="w"> </span>wFullPipeline<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>-params-file<span class="w"> </span>default/fullPipeline_illumina_nanpore_getting_started.yml<span class="w">  </span><span class="se">\</span>
<span class="w">    </span>--s3SignIn<span class="w"> </span><span class="nb">false</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--scratch<span class="w"> </span>/vol/scratch<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--databases<span class="w"> </span>/vol/scratch/databases<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--output<span class="w"> </span>output<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--input.paired.path<span class="w"> </span>test_data/fullPipeline/reads_split.tsv
</code></pre></div>
<p>where</p>
<ul>
<li><code>-work-dir</code> points to a directory that is shared between multiple machines.</li>
<li><code>-profile</code> defines the execution profile that should be used.</li>
<li><code>-entry</code> is the entrypoint of the Toolkit.</li>
<li><code>-params-file</code> sets the parameters file which defines the parameters for all tools. (see input section below)</li>
<li><code>--s3SignIn</code> defines if any S3 login for retrieving inputs is necessary. See the <a href="../configuration/#s3-configuration">S3 configuration</a> section for more information on how to configure the Toolkit for possible S3 input data.</li>
<li><code>--scratch</code> is the directory on the worker node where all intermediate results are saved.</li>
<li><code>--databases</code> is the directory on the worker node where all databases are saved. Already downloaded databases on a shared file system can be configured in the database setting of the corresponding <a href="../database/">database section</a> in the configuration file.</li>
<li><code>--output</code> is the output directory where all results are saved. If you want to know more about which outputs are created, then please refer to the <a href="../modules/introduction/">modules section</a>.</li>
<li><code>--input.paired.path</code> is the path to a tsv file that lists the datasets that should be processed. If you want to provide other parameters, please check the <a href="../pipeline_input/">input section</a>.</li>
</ul>
<div class="admonition note">
<p class="admonition-title">Parameter override</p>
<p>Any parameters defined with a double dash are parameters that override parameters that are already specified in the yaml file.</p>
</div>
<h3 id="input">Input<a class="headerlink" href="#input" title="Permanent link">&para;</a></h3>
<div class="tabbed-set tabbed-alternate" data-tabs="1:3"><input checked="checked" id="__tabbed_1_1" name="__tabbed_1" type="radio" /><input id="__tabbed_1_2" name="__tabbed_1" type="radio" /><input id="__tabbed_1_3" name="__tabbed_1" type="radio" /><div class="tabbed-labels"><label for="__tabbed_1_1">TSV Table</label><label for="__tabbed_1_2">Configuration File</label><label for="__tabbed_1_3">Additional S3 Configuration</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="highlight"><pre><span></span><code>SAMPLE  READS1  READS2
test1   https://openstack.cebitec.uni-bielefeld.de:8080/swift/v1/meta_test/small/read1_1.fq.gz  https://openstack.cebitec.uni-bielefeld.de:8080/swift/v1/meta_test/small/read2_1.fq.gz
test2   https://openstack.cebitec.uni-bielefeld.de:8080/swift/v1/meta_test/small/read1_1.fq.gz  https://openstack.cebitec.uni-bielefeld.de:8080/swift/v1/meta_test/small/read2_1.fq.gz
</code></pre></div>
<p>Must include the columns <code>SAMPLE</code>, <code>READS1</code> and <code>READS2</code>. <code>SAMPLE</code> must contain unique dataset identifiers
without whitespaces or special characters. <code>READS1</code> and <code>READS2</code> are paired reads and can be HTTPS URLs, S3 links or files.</p>
</div>
<div class="tabbed-block">
<div class="highlight"><pre><span></span><code><span class="nt">tempdir</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;tmp&quot;</span>
<span class="nt">summary</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">false</span>
<span class="nt">s3SignIn</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">true</span>
<span class="nt">input</span><span class="p">:</span>
<span class="w">  </span><span class="nt">SRA</span><span class="p">:</span>
<span class="w">    </span><span class="nt">pattern</span><span class="p">:</span>
<span class="w">      </span><span class="nt">ont</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;.+[^(_1|_2)].+fastq.gz$&quot;</span>
<span class="w">      </span><span class="nt">illumina</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;.+(_1|_2).+fastq.gz$&quot;</span>
<span class="w">    </span><span class="nt">S3</span><span class="p">:</span>
<span class="w">      </span><span class="nt">path</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">test_data/fullPipeline/production.tsv</span>
<span class="w">      </span><span class="nt">bucket</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;s3://ftp.era.ebi.ac.uk&quot;</span><span class="w"> </span>
<span class="w">      </span><span class="nt">prefix</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;/vol1/fastq/&quot;</span>
<span class="w">      </span><span class="nt">watch</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">false</span>
<span class="nt">output</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">output</span>
<span class="nt">logDir</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">log</span>
<span class="nt">runid</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1</span>
<span class="nt">databases</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;/vol/scratch/databases&quot;</span>
<span class="nt">publishDirMode</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;symlink&quot;</span>
<span class="nt">logLevel</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1</span>
<span class="nt">scratch</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;/vol/scratch&quot;</span>
<span class="nt">steps</span><span class="p">:</span>
<span class="w">  </span><span class="nt">qc</span><span class="p">:</span>
<span class="w">    </span><span class="nt">fastp</span><span class="p">:</span>
<span class="w">       </span><span class="c1"># For PE data, the adapter sequence auto-detection is disabled by default since the adapters can be trimmed by overlap analysis. However, you can specify --detect_adapter_for_pe to enable it.</span>
<span class="w">       </span><span class="c1"># For PE data, fastp will run a little slower if you specify the sequence adapters or enable adapter auto-detection, but usually result in a slightly cleaner output, since the overlap analysis may fail due to sequencing errors or adapter dimers.</span>
<span class="w">       </span><span class="c1"># -q, --qualified_quality_phred       the quality value that a base is qualified. Default 15 means phred quality &gt;=Q15 is qualified.</span>
<span class="w">       </span><span class="c1"># --cut_front move a sliding window from front (5&#39;) to tail, drop the bases in the window if its mean quality is below cut_mean_quality, stop otherwise.</span>
<span class="w">       </span><span class="c1"># --length_required  reads shorter than length_required will be discarded, default is 15. (int [=15])</span>
<span class="w">       </span><span class="c1"># PE data, the front/tail trimming settings are given with -f, --trim_front1 and -t, --trim_tail1</span>
<span class="w">       </span><span class="nt">additionalParams</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">--detect_adapter_for_pe</span><span class="nv"> </span><span class="s">-q</span><span class="nv"> </span><span class="s">20</span><span class="nv"> </span><span class="s">--cut_front</span><span class="nv"> </span><span class="s">--trim_front1</span><span class="nv"> </span><span class="s">3</span><span class="nv"> </span><span class="s">--cut_tail</span><span class="nv"> </span><span class="s">--trim_tail1</span><span class="nv"> </span><span class="s">3</span><span class="nv"> </span><span class="s">--cut_mean_quality</span><span class="nv"> </span><span class="s">10</span><span class="nv"> </span><span class="s">--length_required</span><span class="nv"> </span><span class="s">50</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">       </span><span class="nt">timeLimit</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;AUTO&quot;</span>
<span class="w">    </span><span class="nt">nonpareil</span><span class="p">:</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">-v</span><span class="nv"> </span><span class="s">10</span><span class="nv"> </span><span class="s">-r</span><span class="nv"> </span><span class="s">1234</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">    </span><span class="nt">kmc</span><span class="p">:</span>
<span class="w">      </span><span class="nt">timeLimit</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;AUTO&quot;</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span>
<span class="w">        </span><span class="c1"># Computes k-mer distribution based on k-mer length 13 and 21</span>
<span class="w">        </span><span class="c1">#  -sm - use strict memory mode (memory limit from -m&lt;n&gt; switch will not be exceeded)</span>
<span class="w">        </span><span class="c1">#  -cs&lt;value&gt; - maximal value of a counter</span>
<span class="w">        </span><span class="nt">count</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">-sm</span><span class="nv"> </span><span class="s">-cs10000</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">        </span><span class="nt">histo</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">-cx50000</span><span class="nv"> </span><span class="s">&quot;</span>

<span class="w">  </span><span class="nt">qcONT</span><span class="p">:</span>
<span class="w">    </span><span class="nt">porechop</span><span class="p">:</span>
<span class="w">       </span><span class="nt">additionalParams</span><span class="p">:</span>
<span class="w">         </span><span class="c1"># Input files are split into chunks because of RAM issues</span>
<span class="w">         </span><span class="nt">chunkSize</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">450000</span>
<span class="w">         </span><span class="nt">porechop</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;&quot;</span>
<span class="w">        </span><span class="c1"># --keep_percent Throw out the worst 10% of reads. This is measured by bp, not by read count. So this option throws out the worst 10% of read bases. </span>
<span class="w">        </span><span class="c1"># </span>
<span class="w">         </span><span class="nt">filtlong</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">--min_length</span><span class="nv"> </span><span class="s">1000</span><span class="nv">  </span><span class="s">--keep_percent</span><span class="nv"> </span><span class="s">90</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">    </span><span class="nt">nanoplot</span><span class="p">:</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;&quot;</span>
<span class="w">  </span><span class="nt">assembly</span><span class="p">:</span>
<span class="w">    </span><span class="nt">megahit</span><span class="p">:</span>
<span class="w">      </span><span class="c1"># --mem-flag 0 to use minimum memory, --mem-flag 1 (default) moderate memory and --mem-flag 2 all memory.</span>
<span class="w">      </span><span class="c1"># meta-sensitive: &#39;--min-count 1 --k-list 21,29,39,49,...,129,141&#39; </span>
<span class="w">      </span><span class="c1"># meta-large:  &#39;--k-min  27  --k-max 127 --k-step 10&#39; (large &amp; complex metagenomes, like soil)</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">--min-contig-len</span><span class="nv"> </span><span class="s">1000</span><span class="nv"> </span><span class="s">--presets</span><span class="nv"> </span><span class="s">meta-sensitive</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">      </span><span class="nt">fastg</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">true</span>
<span class="w">      </span><span class="nt">resources</span><span class="p">:</span>
<span class="w">        </span><span class="nt">RAM</span><span class="p">:</span>
<span class="w">          </span><span class="nt">mode</span><span class="p">:</span><span class="w"> </span><span class="s">&#39;PREDICT&#39;</span>
<span class="w">          </span><span class="nt">predictMinLabel</span><span class="p">:</span><span class="w"> </span><span class="s">&#39;medium&#39;</span>
<span class="w">  </span><span class="nt">assemblyONT</span><span class="p">:</span>
<span class="w">    </span><span class="nt">metaflye</span><span class="p">:</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">-i</span><span class="nv"> </span><span class="s">1</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">      </span><span class="nt">quality</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;AUTO&quot;</span>
<span class="w">  </span><span class="nt">binning</span><span class="p">:</span>
<span class="w">    </span><span class="nt">bwa2</span><span class="p">:</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span><span class="w"> </span>
<span class="w">        </span><span class="nt">bwa2</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">        </span><span class="c1"># samtools flags are used to filter resulting bam file</span>
<span class="w">        </span><span class="nt">samtoolsView</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">-F</span><span class="nv"> </span><span class="s">3584</span><span class="nv"> </span><span class="s">&quot;</span><span class="w"> </span>
<span class="w">    </span><span class="nt">contigsCoverage</span><span class="p">:</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">--min-covered-fraction</span><span class="nv"> </span><span class="s">0</span><span class="nv"> </span><span class="s">--min-read-percent-identity</span><span class="nv"> </span><span class="s">100</span><span class="nv"> </span><span class="s">--min-read-aligned-percent</span><span class="nv"> </span><span class="s">100</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">    </span><span class="nt">genomeCoverage</span><span class="p">:</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">--min-covered-fraction</span><span class="nv"> </span><span class="s">0</span><span class="nv"> </span><span class="s">--min-read-percent-identity</span><span class="nv"> </span><span class="s">100</span><span class="nv"> </span><span class="s">--min-read-aligned-percent</span><span class="nv"> </span><span class="s">100</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">    </span><span class="c1"># Primary binning tool</span>
<span class="w">    </span><span class="nt">metabat</span><span class="p">:</span>
<span class="w">      </span><span class="c1"># Set --seed positive numbers to reproduce the result exactly. Otherwise, random seed will be set each time.</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">--seed</span><span class="nv"> </span><span class="s">234234</span><span class="nv">  </span><span class="s">&quot;</span>
<span class="w">    </span><span class="c1"># Secondary binning tool for use with MAGscot</span>
<span class="w">  </span><span class="nt">binningONT</span><span class="p">:</span>
<span class="w">    </span><span class="nt">minimap</span><span class="p">:</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span><span class="w"> </span>
<span class="w">        </span><span class="nt">minimap</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">        </span><span class="c1"># samtools flags are used to filter resulting bam file</span>
<span class="w">        </span><span class="nt">samtoolsView</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">-F</span><span class="nv"> </span><span class="s">3584</span><span class="nv"> </span><span class="s">&quot;</span><span class="w"> </span>
<span class="w">    </span><span class="nt">contigsCoverage</span><span class="p">:</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">--min-covered-fraction</span><span class="nv"> </span><span class="s">0</span><span class="nv">  </span><span class="s">--min-read-aligned-percent</span><span class="nv"> </span><span class="s">100</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">    </span><span class="nt">genomeCoverage</span><span class="p">:</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">--min-covered-fraction</span><span class="nv"> </span><span class="s">0</span><span class="nv">  </span><span class="s">--min-read-aligned-percent</span><span class="nv"> </span><span class="s">100</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">    </span><span class="nt">metabat</span><span class="p">:</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">--seed</span><span class="nv"> </span><span class="s">234234</span><span class="nv">  </span><span class="s">&quot;</span>
<span class="w">  </span><span class="nt">magAttributes</span><span class="p">:</span>
<span class="w">    </span><span class="c1"># gtdbtk classify_wf</span>
<span class="w">    </span><span class="c1"># --min_af minimum alignment fraction to assign genome to a species cluster (0.5)</span>
<span class="w">    </span><span class="nt">gtdb</span><span class="p">:</span>
<span class="w">      </span><span class="nt">buffer</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1000</span>
<span class="w">      </span><span class="nt">database</span><span class="p">:</span>
<span class="w">        </span><span class="nt">download</span><span class="p">:</span>
<span class="w">          </span><span class="nt">source</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">https://openstack.cebitec.uni-bielefeld.de:8080/databases/gtdbtk_r214_data.tar.gz</span>
<span class="w">          </span><span class="nt">md5sum</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">390e16b3f7b0c4463eb7a3b2149261d9</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">--min_af</span><span class="nv"> </span><span class="s">0.65</span><span class="nv"> </span><span class="s">--scratch_dir</span><span class="nv"> </span><span class="s">.</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">    </span><span class="nt">checkm2</span><span class="p">:</span>
<span class="w">      </span><span class="nt">database</span><span class="p">:</span>
<span class="w">        </span><span class="nt">download</span><span class="p">:</span>
<span class="w">          </span><span class="nt">source</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;https://openstack.cebitec.uni-bielefeld.de:8080/databases/checkm2_v2.tar.gz&quot;</span>
<span class="w">          </span><span class="nt">md5sum</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">a634cb3d31a1f56f2912b74005f25f09</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv">  </span><span class="s">&quot;</span>
<span class="w">  </span><span class="nt">fragmentRecruitment</span><span class="p">:</span>
<span class="w">    </span><span class="nt">mashScreen</span><span class="p">:</span>
<span class="w">      </span><span class="nt">genomes</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">test_data/fragmentRecruitment/mags.tsv</span>
<span class="w">      </span><span class="nt">unzip</span><span class="p">:</span>
<span class="w">        </span><span class="nt">timeLimit</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;AUTO&quot;</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span>
<span class="w">        </span><span class="nt">mashSketch</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">        </span><span class="nt">mashScreen</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">        </span><span class="nt">bwa2</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv">  </span><span class="s">&quot;</span>
<span class="w">        </span><span class="c1"># We demand that matches should be at least 70 percent to restrict the</span>
<span class="w">        </span><span class="c1"># search space. We further restrict all found matches in &#39;coveredBasesCutoff&#39; parameter.</span>
<span class="w">        </span><span class="nt">coverm</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">--min-covered-fraction</span><span class="nv"> </span><span class="s">70</span><span class="nv"> </span><span class="s">--min-read-percent-identity</span><span class="nv"> </span><span class="s">95</span><span class="nv"> </span><span class="s">--min-read-aligned-percent</span><span class="nv"> </span><span class="s">95</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">        </span><span class="nt">covermONT</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">--min-covered-fraction</span><span class="nv"> </span><span class="s">70</span><span class="nv"> </span><span class="s">--min-read-aligned-percent</span><span class="nv"> </span><span class="s">95</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">        </span><span class="nt">minimap</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv">   </span><span class="s">&quot;</span>
<span class="w">        </span><span class="nt">samtoolsViewBwa2</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">-F</span><span class="nv"> </span><span class="s">3584</span><span class="nv"> </span><span class="s">&quot;</span><span class="w"> </span>
<span class="w">        </span><span class="nt">samtoolsViewMinimap</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">&quot;</span><span class="w"> </span>
<span class="w">       </span><span class="c1"># Mash identity should be at least 95%</span>
<span class="w">      </span><span class="nt">mashDistCutoff</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.95</span>
<span class="w">       </span><span class="l l-Scalar l-Scalar-Plain"># percentage of covered genome</span>
<span class="w">      </span><span class="nt">coveredBasesCutoff</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.90</span>
<span class="w">       </span><span class="l l-Scalar l-Scalar-Plain"># Minimum number of k-mer hashes to match</span>
<span class="w">      </span><span class="nt">mashHashCutoff</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">700</span>
<span class="w">    </span><span class="c1"># For final reporting of the covered fraction, the --min-covered-fraction parameter is set to the same value as </span>
<span class="w">    </span><span class="c1"># in the binning and other sections</span>
<span class="w">    </span><span class="nt">genomeCoverage</span><span class="p">:</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">--min-covered-fraction</span><span class="nv"> </span><span class="s">0</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">    </span><span class="nt">contigsCoverage</span><span class="p">:</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">--min-covered-fraction</span><span class="nv"> </span><span class="s">0</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">  </span><span class="nt">dereplication</span><span class="p">:</span>
<span class="w">    </span><span class="nt">bottomUpClustering</span><span class="p">:</span>
<span class="w">      </span><span class="c1"># stricter MIMAG medium quality</span>
<span class="w">      </span><span class="nt">minimumCompleteness</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">50</span>
<span class="w">      </span><span class="nt">maximumContamination</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">5</span>
<span class="w">      </span><span class="nt">ANIBuffer</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">20</span>
<span class="w">      </span><span class="nt">mashBuffer</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">2000</span>
<span class="w">      </span><span class="nt">method</span><span class="p">:</span><span class="w"> </span><span class="s">&#39;ANI&#39;</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span>
<span class="w">        </span><span class="nt">mash_sketch</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;&quot;</span>
<span class="w">        </span><span class="nt">mash_dist</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;&quot;</span>
<span class="w">        </span><span class="c1"># cluster cutoff</span>
<span class="w">        </span><span class="nt">cluster</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">-c</span><span class="nv"> </span><span class="s">0.05</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">        </span><span class="nt">pyani</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">-m</span><span class="nv"> </span><span class="s">ANIb</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">        </span><span class="nt">representativeAniCutoff</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.95</span>
<span class="w">  </span><span class="nt">readMapping</span><span class="p">:</span>
<span class="w">    </span><span class="nt">bwa2</span><span class="p">:</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span>
<span class="w">        </span><span class="nt">bwa2_index</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;&quot;</span>
<span class="w">        </span><span class="nt">bwa2_mem</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;&quot;</span>
<span class="w">     </span><span class="c1"># This module produces two abundance tables.</span>
<span class="w">     </span><span class="c1"># One table is based on relative abundance and the second one on the trimmed mean.</span>
<span class="w">     </span><span class="c1"># Just using relative abundance makes it difficult to tell if a genome is part of a dataset.</span>
<span class="w">     </span><span class="c1"># Thats why it makes sense to set at leat a low min covered fraction parameter.</span>
<span class="w">    </span><span class="nt">coverm</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">--min-covered-fraction</span><span class="nv"> </span><span class="s">80</span><span class="nv">  </span><span class="s">--min-read-percent-identity</span><span class="nv"> </span><span class="s">95</span><span class="nv"> </span><span class="s">--min-read-aligned-percent</span><span class="nv"> </span><span class="s">95</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">    </span><span class="nt">covermONT</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">--min-covered-fraction</span><span class="nv"> </span><span class="s">80</span><span class="nv">  </span><span class="s">--min-read-aligned-percent</span><span class="nv"> </span><span class="s">95</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">    </span><span class="nt">minimap</span><span class="p">:</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span>
<span class="w">        </span><span class="nt">minimap_index</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;&quot;</span>
<span class="w">        </span><span class="nt">minimap</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;&quot;</span>
<span class="w">  </span><span class="nt">cooccurrence</span><span class="p">:</span>
<span class="w">    </span><span class="nt">inference</span><span class="p">:</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span>
<span class="w">        </span><span class="nt">method</span><span class="p">:</span><span class="w"> </span><span class="s">&#39;correlation&#39;</span>
<span class="w">        </span><span class="nt">rscript</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">--mincovthreshold</span><span class="nv"> </span><span class="s">0.9</span><span class="nv"> </span><span class="s">--maxzero</span><span class="nv"> </span><span class="s">60</span><span class="nv"> </span><span class="s">--minweight</span><span class="nv"> </span><span class="s">0.4</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">        </span><span class="nt">timeLimit</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;AUTO&quot;</span>
<span class="w">    </span><span class="nt">metabolicAnnotation</span><span class="p">:</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span>
<span class="w">        </span><span class="nt">metabolicEdgeBatches</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">5</span>
<span class="w">        </span><span class="nt">metabolicEdgeReplicates</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">10</span>
<span class="w">        </span><span class="nt">smetana</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">--flavor</span><span class="nv"> </span><span class="s">bigg</span><span class="nv"> </span><span class="s">--molweight</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">  </span><span class="nt">plasmid</span><span class="p">:</span>
<span class="w">    </span><span class="nt">SCAPP</span><span class="p">:</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span>
<span class="w">        </span><span class="nt">bwa2</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv">  </span><span class="s">&quot;</span>
<span class="w">        </span><span class="nt">SCAPP</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv">  </span><span class="s">&quot;</span>
<span class="w">        </span><span class="nt">coverm</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">--min-covered-fraction</span><span class="nv"> </span><span class="s">0</span><span class="nv"> </span><span class="s">--min-read-percent-identity</span><span class="nv"> </span><span class="s">100</span><span class="nv"> </span><span class="s">--min-read-aligned-percent</span><span class="nv"> </span><span class="s">100</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">        </span><span class="nt">covermONT</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">--min-covered-fraction</span><span class="nv"> </span><span class="s">0</span><span class="nv"> </span><span class="s">--min-read-aligned-percent</span><span class="nv"> </span><span class="s">100</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">        </span><span class="nt">minimap</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">        </span><span class="nt">samtoolsViewBwa2</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">-F</span><span class="nv"> </span><span class="s">3584</span><span class="nv"> </span><span class="s">&quot;</span><span class="w"> </span>
<span class="w">        </span><span class="nt">samtoolsViewMinimap</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">&quot;</span><span class="w"> </span>
<span class="w">    </span><span class="nt">ViralVerifyPlasmid</span><span class="p">:</span>
<span class="w">      </span><span class="nt">filter</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">true</span>
<span class="w">      </span><span class="c1"># Select sequences that are labeled as uncertain or plasmid id ViralVerifies output</span>
<span class="w">      </span><span class="nt">filterString</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;Uncertain</span><span class="nv"> </span><span class="s">-</span><span class="nv"> </span><span class="s">plasmid</span><span class="nv"> </span><span class="s">or</span><span class="nv"> </span><span class="s">chromosomal|Plasmid&quot;</span>
<span class="w">      </span><span class="nt">database</span><span class="p">:</span>
<span class="w">        </span><span class="nt">download</span><span class="p">:</span>
<span class="w">          </span><span class="nt">source</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">https://openstack.cebitec.uni-bielefeld.de:8080/databases/pfam-A_35.0.hmm.gz</span>
<span class="w">          </span><span class="nt">md5sum</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">c80b75bd48ec41760bbca19c70616e36</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">--thr</span><span class="nv"> </span><span class="s">7</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">    </span><span class="nt">PlasClass</span><span class="p">:</span>
<span class="w">      </span><span class="nt">filter</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">true</span>
<span class="w">      </span><span class="c1"># A cutoff of 0.9 for longer (&gt; 2 kb) sequences should pretty much yield only plasmids.</span>
<span class="w">      </span><span class="nt">threshold</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.9</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv">   </span><span class="s">&quot;</span>
<span class="w">    </span><span class="nt">Filter</span><span class="p">:</span>
<span class="w">      </span><span class="nt">method</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;OR&quot;</span>
<span class="w">      </span><span class="nt">minLength</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1000</span><span class="w"> </span>
<span class="w">    </span><span class="nt">PLSDB</span><span class="p">:</span>
<span class="w">      </span><span class="nt">sharedKmerThreshold</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0</span>
<span class="w">      </span><span class="nt">database</span><span class="p">:</span>
<span class="w">        </span><span class="nt">download</span><span class="p">:</span>
<span class="w">          </span><span class="nt">source</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">https://openstack.cebitec.uni-bielefeld.de:8080/databases/plasmids_plsdb_20220929.tar.bz2</span>
<span class="w">          </span><span class="nt">md5sum</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">13c1078e6cd6a46e3f508c24ca07cc18</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span>
<span class="w">     </span><span class="c1"># Parameters according to plsdb paper</span>
<span class="w">     </span><span class="c1"># -s: Sketch size</span>
<span class="w">     </span><span class="c1"># -k: k-mer size</span>
<span class="w">     </span><span class="c1"># -S: Seed to provide to the hash function</span>
<span class="w">     </span><span class="c1"># -d: Maximum distance to report</span>
<span class="w">     </span><span class="c1"># -v: Maximum p-value to report</span>
<span class="w">        </span><span class="nt">mashSketch</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">-S</span><span class="nv"> </span><span class="s">42</span><span class="nv"> </span><span class="s">-k</span><span class="nv"> </span><span class="s">21</span><span class="nv"> </span><span class="s">-s</span><span class="nv"> </span><span class="s">1000</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">        </span><span class="nt">mashDist</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">-v</span><span class="nv"> </span><span class="s">0.1</span><span class="nv"> </span><span class="s">-d</span><span class="nv"> </span><span class="s">0.1</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">    </span><span class="nt">MobTyper</span><span class="p">:</span>
<span class="w">      </span><span class="nt">minLength</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1000</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">--min_length</span><span class="nv"> </span><span class="s">1000</span><span class="nv">  </span><span class="s">&quot;</span>
<span class="w">      </span><span class="nt">database</span><span class="p">:</span>
<span class="w">        </span><span class="nt">download</span><span class="p">:</span>
<span class="w">          </span><span class="nt">source</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">https://openstack.cebitec.uni-bielefeld.de:8080/databases/mob_20220929.gz</span>
<span class="w">          </span><span class="nt">md5sum</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">21fcaf9c3754a985d1d6875939d71e28</span>
<span class="w">  </span><span class="nt">annotation</span><span class="p">:</span>
<span class="w">    </span><span class="nt">prokka</span><span class="p">:</span>
<span class="w">      </span><span class="nt">defaultKingdom</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">false</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">--mincontiglen</span><span class="nv"> </span><span class="s">500</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">    </span><span class="nt">mmseqs2</span><span class="p">:</span>
<span class="w">      </span><span class="nt">chunkSize</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">20000</span>
<span class="w">      </span><span class="nt">kegg</span><span class="p">:</span>
<span class="w">        </span><span class="nt">additionalParams</span><span class="p">:</span>
<span class="w">          </span><span class="nt">search </span><span class="p">:</span><span class="w"> </span><span class="s">&#39;</span><span class="nv"> </span><span class="s">--max-seqs</span><span class="nv"> </span><span class="s">300</span><span class="nv"> </span><span class="s">--max-accept</span><span class="nv"> </span><span class="s">50</span><span class="nv"> </span><span class="s">-c</span><span class="nv"> </span><span class="s">0</span><span class="nv"> </span><span class="s">--cov-mode</span><span class="nv"> </span><span class="s">2</span><span class="nv"> </span><span class="s">--start-sens</span><span class="nv"> </span><span class="s">4</span><span class="nv"> </span><span class="s">--sens-steps</span><span class="nv"> </span><span class="s">1</span><span class="nv"> </span><span class="s">-s</span><span class="nv"> </span><span class="s">6</span><span class="nv"> </span><span class="s">--num-iterations</span><span class="nv"> </span><span class="s">2</span><span class="nv"> </span><span class="s">-e</span><span class="nv"> </span><span class="s">0.001</span><span class="nv"> </span><span class="s">--e-profile</span><span class="nv"> </span><span class="s">0.01</span><span class="nv"> </span><span class="s">--db-load-mode</span><span class="nv"> </span><span class="s">3</span><span class="nv"> </span><span class="s">&#39;</span>
<span class="w">          </span><span class="nt">additionalColumns</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;&quot;</span>
<span class="w">        </span><span class="nt">database</span><span class="p">:</span>
<span class="w">          </span><span class="nt">download</span><span class="p">:</span>
<span class="w">            </span><span class="nt">source</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">s3://databases_internal/kegg-mirror-2021-01_mmseqs.tar.zst</span>
<span class="w">            </span><span class="nt">md5sum</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0d20db97b3e7ee6571ca1fd5ad3a87f1</span>
<span class="w">            </span><span class="nt">s5cmd</span><span class="p">:</span>
<span class="w">              </span><span class="nt">params</span><span class="p">:</span><span class="w"> </span><span class="s">&#39;--retry-count</span><span class="nv"> </span><span class="s">30</span><span class="nv"> </span><span class="s">--no-verify-ssl</span><span class="nv"> </span><span class="s">--endpoint-url</span><span class="nv"> </span><span class="s">https://openstack.cebitec.uni-bielefeld.de:8080&#39;</span>
<span class="w">      </span><span class="nt">vfdb</span><span class="p">:</span>
<span class="w">        </span><span class="nt">additionalParams</span><span class="p">:</span>
<span class="w">          </span><span class="nt">search </span><span class="p">:</span><span class="w"> </span><span class="s">&#39;</span><span class="nv"> </span><span class="s">--max-seqs</span><span class="nv"> </span><span class="s">300</span><span class="nv"> </span><span class="s">--max-accept</span><span class="nv"> </span><span class="s">50</span><span class="nv"> </span><span class="s">-c</span><span class="nv"> </span><span class="s">0</span><span class="nv"> </span><span class="s">--cov-mode</span><span class="nv"> </span><span class="s">2</span><span class="nv"> </span><span class="s">--start-sens</span><span class="nv"> </span><span class="s">4</span><span class="nv"> </span><span class="s">--sens-steps</span><span class="nv"> </span><span class="s">1</span><span class="nv"> </span><span class="s">-s</span><span class="nv"> </span><span class="s">6</span><span class="nv"> </span><span class="s">--num-iterations</span><span class="nv"> </span><span class="s">2</span><span class="nv"> </span><span class="s">-e</span><span class="nv"> </span><span class="s">0.001</span><span class="nv"> </span><span class="s">--e-profile</span><span class="nv"> </span><span class="s">0.01</span><span class="nv"> </span><span class="s">--db-load-mode</span><span class="nv"> </span><span class="s">3</span><span class="nv"> </span><span class="s">&#39;</span>
<span class="w">          </span><span class="nt">additionalColumns</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;&quot;</span>
<span class="w">        </span><span class="nt">database</span><span class="p">:</span>
<span class="w">          </span><span class="nt">download</span><span class="p">:</span>
<span class="w">            </span><span class="nt">source</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">https://openstack.cebitec.uni-bielefeld.de:8080/databases/vfdb_full_2022_07_29.tar.zst</span>
<span class="w">            </span><span class="nt">md5sum</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">7e32aaed112d6e056fb8764b637bf49e</span>
<span class="w">      </span><span class="nt">bacmet20_experimental</span><span class="p">:</span>
<span class="w">        </span><span class="nt">additionalParams</span><span class="p">:</span>
<span class="w">          </span><span class="nt">search </span><span class="p">:</span><span class="w"> </span><span class="s">&#39;</span><span class="nv"> </span><span class="s">--max-seqs</span><span class="nv"> </span><span class="s">300</span><span class="nv"> </span><span class="s">--max-accept</span><span class="nv"> </span><span class="s">50</span><span class="nv"> </span><span class="s">-c</span><span class="nv"> </span><span class="s">0</span><span class="nv"> </span><span class="s">--cov-mode</span><span class="nv"> </span><span class="s">2</span><span class="nv"> </span><span class="s">--start-sens</span><span class="nv"> </span><span class="s">4</span><span class="nv"> </span><span class="s">--sens-steps</span><span class="nv"> </span><span class="s">1</span><span class="nv"> </span><span class="s">-s</span><span class="nv"> </span><span class="s">6</span><span class="nv"> </span><span class="s">--num-iterations</span><span class="nv"> </span><span class="s">2</span><span class="nv"> </span><span class="s">-e</span><span class="nv"> </span><span class="s">0.001</span><span class="nv"> </span><span class="s">--e-profile</span><span class="nv"> </span><span class="s">0.01</span><span class="nv"> </span><span class="s">--db-load-mode</span><span class="nv"> </span><span class="s">3</span><span class="nv"> </span><span class="s">&#39;</span>
<span class="w">          </span><span class="nt">additionalColumns</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;&quot;</span>
<span class="w">        </span><span class="nt">database</span><span class="p">:</span>
<span class="w">          </span><span class="nt">download</span><span class="p">:</span>
<span class="w">            </span><span class="nt">source</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">https://openstack.cebitec.uni-bielefeld.de:8080/databases/bacmet20_experimental.tar.zst</span>
<span class="w">            </span><span class="nt">md5sum</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">57a6d328486f0acd63f7e984f739e8fe</span>
<span class="w">      </span><span class="nt">bacmet20_predicted</span><span class="p">:</span>
<span class="w">        </span><span class="nt">additionalParams</span><span class="p">:</span>
<span class="w">          </span><span class="nt">search </span><span class="p">:</span><span class="w"> </span><span class="s">&#39;</span><span class="nv"> </span><span class="s">--max-seqs</span><span class="nv"> </span><span class="s">300</span><span class="nv"> </span><span class="s">--max-accept</span><span class="nv"> </span><span class="s">50</span><span class="nv"> </span><span class="s">-c</span><span class="nv"> </span><span class="s">0</span><span class="nv"> </span><span class="s">--cov-mode</span><span class="nv"> </span><span class="s">2</span><span class="nv"> </span><span class="s">--start-sens</span><span class="nv"> </span><span class="s">4</span><span class="nv"> </span><span class="s">--sens-steps</span><span class="nv"> </span><span class="s">1</span><span class="nv"> </span><span class="s">-s</span><span class="nv"> </span><span class="s">6</span><span class="nv"> </span><span class="s">--num-iterations</span><span class="nv"> </span><span class="s">2</span><span class="nv"> </span><span class="s">-e</span><span class="nv"> </span><span class="s">0.001</span><span class="nv"> </span><span class="s">--e-profile</span><span class="nv"> </span><span class="s">0.01</span><span class="nv"> </span><span class="s">--db-load-mode</span><span class="nv"> </span><span class="s">3</span><span class="nv"> </span><span class="s">&#39;</span>
<span class="w">          </span><span class="nt">additionalColumns</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;&quot;</span>
<span class="w">        </span><span class="nt">database</span><span class="p">:</span>
<span class="w">          </span><span class="nt">download</span><span class="p">:</span>
<span class="w">            </span><span class="nt">source</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">https://openstack.cebitec.uni-bielefeld.de:8080/databases/bacmet20_predicted.tar.zst</span>
<span class="w">            </span><span class="nt">md5sum</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">55902401a765fc460c09994d839d9b64</span>
<span class="w">      </span><span class="nt">uniref90</span><span class="p">:</span>
<span class="w">        </span><span class="nt">additionalParams</span><span class="p">:</span>
<span class="w">          </span><span class="nt">search </span><span class="p">:</span><span class="w"> </span><span class="s">&#39;</span><span class="nv"> </span><span class="s">--max-seqs</span><span class="nv"> </span><span class="s">300</span><span class="nv"> </span><span class="s">--max-accept</span><span class="nv"> </span><span class="s">50</span><span class="nv"> </span><span class="s">-c</span><span class="nv"> </span><span class="s">0</span><span class="nv"> </span><span class="s">--cov-mode</span><span class="nv"> </span><span class="s">2</span><span class="nv"> </span><span class="s">--start-sens</span><span class="nv"> </span><span class="s">4</span><span class="nv"> </span><span class="s">--sens-steps</span><span class="nv"> </span><span class="s">1</span><span class="nv"> </span><span class="s">-s</span><span class="nv"> </span><span class="s">6</span><span class="nv"> </span><span class="s">--num-iterations</span><span class="nv"> </span><span class="s">2</span><span class="nv"> </span><span class="s">-e</span><span class="nv"> </span><span class="s">0.001</span><span class="nv"> </span><span class="s">--e-profile</span><span class="nv"> </span><span class="s">0.01</span><span class="nv"> </span><span class="s">--db-load-mode</span><span class="nv"> </span><span class="s">3</span><span class="nv"> </span><span class="s">&#39;</span>
<span class="w">          </span><span class="nt">additionalColumns</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;&quot;</span>
<span class="w">        </span><span class="nt">database</span><span class="p">:</span>
<span class="w">          </span><span class="nt">download</span><span class="p">:</span>
<span class="w">            </span><span class="nt">source</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">https://openstack.cebitec.uni-bielefeld.de:8080/databases/uniref90_20231108_mmseqs.tar.zst</span>
<span class="w">            </span><span class="nt">md5sum</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">313f2c031361091af2d5f3c6f6f46013</span>
<span class="w">    </span><span class="nt">rgi</span><span class="p">:</span>
<span class="w">      </span><span class="c1"># --include_loose includes matches of more distant homologs of AMR genes which may also report spurious partial matches</span>
<span class="w">      </span><span class="c1"># --include_nudge Partial ORFs may do not pass curated bitscore cut-offs or novel samples may contain divergent alleles, so nudging </span>
<span class="w">      </span><span class="c1">#                 95% identity Loose matches to Strict matches may aid resistome annotation </span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">--include_loose</span><span class="nv"> </span><span class="s">--include_nudge</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">      </span><span class="nt">database</span><span class="p">:</span>
<span class="w">        </span><span class="nt">download</span><span class="p">:</span>
<span class="w">          </span><span class="nt">source</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">https://openstack.cebitec.uni-bielefeld.de:8080/databases/card_20221209.tar.bz2</span>
<span class="w">          </span><span class="nt">md5sum</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">d7e627221a1d4584d1c1795cda774cdb</span>
<span class="w">    </span><span class="nt">mmseqs2_taxonomy</span><span class="p">:</span>
<span class="w">      </span><span class="c1"># Run taxonomy classification on MAGs and unbinable contigs or just the later</span>
<span class="w">      </span><span class="nt">runOnMAGs</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">true</span>
<span class="w">      </span><span class="nt">gtdb</span><span class="p">:</span>
<span class="w">        </span><span class="c1"># High sensitivity mode parameters only work for quality MAGs, for unbinable contigs use default parameters</span>
<span class="w">        </span><span class="nt">params</span><span class="p">:</span><span class="w"> </span><span class="s">&#39;</span><span class="nv"> </span><span class="s">--lca-ranks</span><span class="nv"> </span><span class="s">superkingdom,phylum,class,order,family,genus,species,subspecies</span><span class="nv"> </span><span class="s">--max-seqs</span><span class="nv"> </span><span class="s">300</span><span class="nv"> </span><span class="s">--max-accept</span><span class="nv"> </span><span class="s">50</span><span class="nv"> </span><span class="s">--cov-mode</span><span class="nv"> </span><span class="s">2</span><span class="nv"> </span><span class="s">-e</span><span class="nv"> </span><span class="s">0.001</span><span class="nv"> </span><span class="s">--e-profile</span><span class="nv"> </span><span class="s">0.01</span><span class="nv"> </span><span class="s">&#39;</span>
<span class="w">        </span><span class="c1"># Load database into memory to speed up classification, only works if --db-load-mode 3 is set in params</span>
<span class="w">        </span><span class="nt">ramMode</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">false</span>
<span class="w">        </span><span class="nt">initialMaxSensitivity</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">6</span>
<span class="w">        </span><span class="nt">database</span><span class="p">:</span>
<span class="w">          </span><span class="nt">download</span><span class="p">:</span>
<span class="w">            </span><span class="nt">source</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">https://openstack.cebitec.uni-bielefeld.de:8080/databases/gtdb_r214_1_mmseqs.tar.gz</span>
<span class="w">            </span><span class="nt">md5sum</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">3c8f12c5c2dc55841a14dd30a0a4c718</span>
<span class="w">    </span><span class="nt">keggFromMMseqs2</span><span class="p">:</span>
<span class="w">      </span><span class="nt">database</span><span class="p">:</span>
<span class="w">        </span><span class="nt">download</span><span class="p">:</span>
<span class="w">          </span><span class="nt">source</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">s3://databases_internal/kegg-links-mirror-2021-01.tar.gz</span>
<span class="w">          </span><span class="nt">md5sum</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">69538151ecf4128f4cb1e0ac21ff55b3</span>
<span class="w">          </span><span class="nt">s5cmd</span><span class="p">:</span>
<span class="w">              </span><span class="nt">params</span><span class="p">:</span><span class="w"> </span><span class="s">&#39;--retry-count</span><span class="nv"> </span><span class="s">30</span><span class="nv"> </span><span class="s">--no-verify-ssl</span><span class="nv"> </span><span class="s">--endpoint-url</span><span class="nv"> </span><span class="s">https://openstack.cebitec.uni-bielefeld.de:8080&#39;</span>
<span class="w">  </span><span class="nt">metabolomics</span><span class="p">:</span>
<span class="w">     </span><span class="nt">filter</span><span class="p">:</span>
<span class="w">       </span><span class="nt">minCompleteness</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">50</span>
<span class="w">       </span><span class="nt">maxContamination</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">5</span>
<span class="w">     </span><span class="nt">carveme</span><span class="p">:</span>
<span class="w">       </span><span class="nt">additionalParams</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">--solver</span><span class="nv"> </span><span class="s">scip</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">     </span><span class="nt">smetana</span><span class="p">:</span>
<span class="w">       </span><span class="nt">beforeProcessScript</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;&quot;</span>
<span class="w">       </span><span class="nt">global</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">false</span>
<span class="w">       </span><span class="nt">detailed</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">false</span>
<span class="w">       </span><span class="nt">additionalParams</span><span class="p">:</span>
<span class="w">         </span><span class="nt">detailed</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;&quot;</span>
<span class="w">         </span><span class="nt">global</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;&quot;</span>
<span class="w">     </span><span class="nt">memote</span><span class="p">:</span>
<span class="w">       </span><span class="nt">beforeProcessScript</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;&quot;</span>
<span class="w">       </span><span class="nt">additionalParams</span><span class="p">:</span>
<span class="w">         </span><span class="nt">run</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;&quot;</span>
<span class="w">         </span><span class="nt">report</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;&quot;</span>
<span class="w">  </span><span class="nt">export</span><span class="p">:</span>
<span class="w">    </span><span class="nt">emgb</span><span class="p">:</span>
<span class="w">      </span><span class="nt">titles</span><span class="p">:</span>
<span class="w">        </span><span class="nt">database</span><span class="p">:</span>
<span class="w">          </span><span class="nt">download</span><span class="p">:</span>
<span class="w">            </span><span class="nt">source</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;https://openstack.cebitec.uni-bielefeld.de:8080/databases/uniref90.titles.tsv.gz&quot;</span>
<span class="w">            </span><span class="nt">md5sum</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">aaf1dd9021243def8e6c4e438b4b3669</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span>
<span class="w">              </span><span class="nt">blastDB</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;&quot;</span>
<span class="w">              </span><span class="nt">taxonomyDB</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;&quot;</span>
<span class="w">      </span><span class="nt">kegg</span><span class="p">:</span>
<span class="w">        </span><span class="nt">database</span><span class="p">:</span>
<span class="w">          </span><span class="nt">download</span><span class="p">:</span>
<span class="w">            </span><span class="nt">source</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">s3://databases_internal/annotatedgenes2json_db_kegg-mirror-2022-12.tar.zst</span>
<span class="w">            </span><span class="nt">md5sum</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">262dab8ca564fbc1f27500c22b5bc47b</span>
<span class="w">            </span><span class="nt">s5cmd</span><span class="p">:</span>
<span class="w">              </span><span class="nt">params</span><span class="p">:</span><span class="w"> </span><span class="s">&#39;--retry-count</span><span class="nv"> </span><span class="s">30</span><span class="nv"> </span><span class="s">--no-verify-ssl</span><span class="nv"> </span><span class="s">--endpoint-url</span><span class="nv"> </span><span class="s">https://openstack.cebitec.uni-bielefeld.de:8080&#39;</span>
<span class="nt">resources</span><span class="p">:</span>
<span class="w">  </span><span class="nt">highmemLarge</span><span class="p">:</span>
<span class="w">    </span><span class="nt">cpus</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">28</span>
<span class="w">    </span><span class="nt">memory</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">230</span>
<span class="w">  </span><span class="nt">highmemMedium</span><span class="p">:</span>
<span class="w">    </span><span class="nt">cpus</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">14</span>
<span class="w">    </span><span class="nt">memory</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">113</span>
<span class="w">  </span><span class="nt">large</span><span class="p">:</span>
<span class="w">    </span><span class="nt">cpus</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">28</span>
<span class="w">    </span><span class="nt">memory</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">58</span>
<span class="w">  </span><span class="nt">medium</span><span class="p">:</span>
<span class="w">    </span><span class="nt">cpus</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">14</span>
<span class="w">    </span><span class="nt">memory</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">29</span>
<span class="w">  </span><span class="nt">small</span><span class="p">:</span>
<span class="w">    </span><span class="nt">cpus</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">7</span>
<span class="w">    </span><span class="nt">memory</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">14</span>
<span class="w">  </span><span class="nt">tiny</span><span class="p">:</span>
<span class="w">    </span><span class="nt">cpus</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1</span>
<span class="w">    </span><span class="nt">memory</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1</span>
</code></pre></div>
</div>
<div class="tabbed-block">
<p>Nextflow usually stores downloaded files in the work directory. If enough scratch disc space is available on the worker nodes then this can be prevented by specifying
s3 links in the input tsv file and <code>download</code> parameter in the qc part of the input yaml.</p>
<div class="admonition note">
<p><strong>Currently the S3 links must point to files that are publicly available.</strong></p>
</div>
<p>S3 TSV Links:
<div class="highlight"><pre><span></span><code>SAMPLE  READS1  READS2
test1   s3://meta_test/small/read1_1.fq.gz  s3://meta_test/small/read2_1.fq.gz
test2   s3://meta_test/small/read1_1.fq.gz  s3://meta_test/small/read2_1.fq.gz
</code></pre></div></p>
<p>Example Input YAML with the <code>download</code> parameter in the qc part.
<div class="highlight"><pre><span></span><code><span class="nt">tempdir</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;tmp&quot;</span>
<span class="nt">summary</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">true</span>
<span class="nt">s3SignIn</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">true</span>
<span class="nt">input</span><span class="p">:</span>
<span class="w">  </span><span class="nt">paired</span><span class="p">:</span><span class="w"> </span>
<span class="w">    </span><span class="nt">path</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;test_data/fullPipeline/reads_split_s3.tsv&quot;</span>
<span class="w">    </span><span class="nt">watch</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">false</span>
<span class="nt">output</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;output&quot;</span>
<span class="nt">logDir</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">log</span>
<span class="nt">runid</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1</span>
<span class="nt">logLevel</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1</span>
<span class="nt">scratch</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;/vol/scratch&quot;</span>
<span class="nt">publishDirMode</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;symlink&quot;</span>
<span class="nt">steps</span><span class="p">:</span>
<span class="w">  </span><span class="nt">qc</span><span class="p">:</span>
<span class="w">    </span><span class="nt">interleaved</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">false</span>
<span class="w">    </span><span class="nt">fastp</span><span class="p">:</span>
<span class="w">      </span><span class="nt">download</span><span class="p">:</span>
<span class="w">        </span><span class="nt">s5cmdParams</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">--retry-count</span><span class="nv"> </span><span class="s">30</span><span class="nv"> </span><span class="s">--no-verify-ssl</span><span class="nv"> </span><span class="s">--no-sign-request</span><span class="nv"> </span><span class="s">--endpoint-url</span><span class="nv"> </span><span class="s">https://openstack.cebitec.uni-bielefeld.de:8080</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">      </span><span class="nt">timeLimit</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;AUTO&quot;</span>
<span class="nt">resources</span><span class="p">:</span>
<span class="w">  </span><span class="nt">highmemLarge</span><span class="p">:</span>
<span class="w">    </span><span class="nt">cpus</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">28</span>
<span class="w">    </span><span class="nt">memory</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">230</span>
<span class="w">  </span><span class="nt">highmemMedium</span><span class="p">:</span>
<span class="w">    </span><span class="nt">cpus</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">14</span>
<span class="w">    </span><span class="nt">memory</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">113</span>
<span class="w">  </span><span class="nt">large</span><span class="p">:</span>
<span class="w">    </span><span class="nt">cpus</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">28</span>
<span class="w">    </span><span class="nt">memory</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">58</span>
<span class="w">  </span><span class="nt">medium</span><span class="p">:</span>
<span class="w">    </span><span class="nt">cpus</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">14</span>
<span class="w">    </span><span class="nt">memory</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">29</span>
<span class="w">  </span><span class="nt">small</span><span class="p">:</span>
<span class="w">    </span><span class="nt">cpus</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">7</span>
<span class="w">    </span><span class="nt">memory</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">14</span>
<span class="w">  </span><span class="nt">tiny</span><span class="p">:</span>
<span class="w">    </span><span class="nt">cpus</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1</span>
<span class="w">    </span><span class="nt">memory</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1</span>
</code></pre></div></p>
</div>
</div>
</div>
<h3 id="output">Output<a class="headerlink" href="#output" title="Permanent link">&para;</a></h3>
<p>The produced output can be inspected on on the <a href="../modules/introduction/">modules page</a>.</p>
<h2 id="further-reading">Further Reading<a class="headerlink" href="#further-reading" title="Permanent link">&para;</a></h2>
<ul>
<li>
<p>Pipeline Configuration: If you want to configure and optimize the Toolkit for your data or your infrastructure you can continue with the <a href="../configuration/">configuration</a> section.</p>
</li>
<li>
<p>In case you want to import the output to EMGB, please go on to the <a href="../emgb/">EMGB</a> part.</p>
</li>
<li>
<p>If you want to add databases or more the pre-configured ones, you can read <a href="../database/">here</a> how to do this.</p>
</li>
<li>
<p>You might want to adjust the <a href="../configuration/#configuration-of-computational-resources-used-for-pipeline-runs">resource requirements</a> of the Toolkit to your infrastructure.</p>
</li>
</ul>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "..", "features": ["navigation.tabs", "navigation.tabs.sticky", "content.code.copy"], "search": "../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"provider": "mike"}}</script>
    
    
      <script src="../assets/javascripts/bundle.bd41221c.min.js"></script>
      
    
  </body>
</html>