
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../assigning_lineage_and_function/">
      
      
        <link rel="next" href="../../../developer_guidelines/">
      
      
      <link rel="icon" href="../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.5.14">
    
    
      
        <title>Dereplication, and Coverage Inspection of MAGs - Metagenomics-Toolkit</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.10ba22f1.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="blue-grey" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#metagenomics-toolkit-execution" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      <div data-md-color-scheme="default" data-md-component="outdated" hidden>
        
          <aside class="md-banner md-banner--warning">
            <div class="md-banner__inner md-grid md-typeset">
              
  You're not viewing the latest version.
    <a href="../../../..">
	        <strong>Click here to go to latest.</strong>
    </a>

            </div>
            <script>var el=document.querySelector("[data-md-component=outdated]"),outdated=__md_get("__outdated",sessionStorage);!0===outdated&&el&&(el.hidden=!1)</script>
          </aside>
        
      </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="Metagenomics-Toolkit" class="md-header__button md-logo" aria-label="Metagenomics-Toolkit" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Metagenomics-Toolkit
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Dereplication, and Coverage Inspection of MAGs
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="blue-grey" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="blue-grey" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/metagenomics/metagenomics-tk" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    Metagenomics-Tk
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../.." class="md-tabs__link">
        
  
    
  
  Introduction

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../../quickstart/" class="md-tabs__link">
        
  
    
  
  Quickstart

      </a>
    </li>
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../../overview/" class="md-tabs__link">
          
  
  Getting Started

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../../configuration/" class="md-tabs__link">
          
  
  Configuration

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../../modules/introduction/" class="md-tabs__link">
          
  
  Modules

        </a>
      </li>
    
  

      
        
  
  
    
  
  
    
    
      
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../../tutorial1/introduction/" class="md-tabs__link">
          
  
  Tutorials

        </a>
      </li>
    
  

    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../../developer_guidelines/" class="md-tabs__link">
          
  
  Developer

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="Metagenomics-Toolkit" class="md-nav__button md-logo" aria-label="Metagenomics-Toolkit" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    Metagenomics-Toolkit
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/metagenomics/metagenomics-tk" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    Metagenomics-Tk
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../quickstart/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Quickstart
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Getting Started
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Getting Started
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../overview/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../full_pipeline/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Full Pipeline
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../aggregation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Aggregation
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Configuration
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Configuration
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../configuration/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Global Pipeline Configuration
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../pipeline_input/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Pipeline Input Configuration
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../database/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Database Configuration
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Modules
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Modules
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/introduction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/qualityControl/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Quality Control
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/annotation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Annotation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/assembly/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Assembly
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/cooccurrence/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Cooccurrence
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/dereplication/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Dereplication
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/fragment_recruitment/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Run Fragment Recruitment
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/magAttributes/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    MagAttributes
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/plasmids/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Plasmids
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/metabolomics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Metabolomics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/readMapping/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Read Mapping
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/export/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Exploratory Metagenome Browser (EMGB)
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
      
        
        
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6" checked>
        
          
          <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Tutorials
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_6">
            <span class="md-nav__icon md-icon"></span>
            Tutorials
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6_1" >
        
          
          <label class="md-nav__link" for="__nav_6_1" id="__nav_6_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Tutorial (Basics)
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_6_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6_1">
            <span class="md-nav__icon md-icon"></span>
            Tutorial (Basics)
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tutorial1/introduction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tutorial1/quality_control/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Quality Control
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tutorial1/assembly/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Assembly
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tutorial1/assembly_evaluation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Assembly Evaluation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tutorial1/binning/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Binning
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tutorial1/bin_quality/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Assessing Bin Quality
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tutorial1/classification/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Classification
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tutorial1/annotation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Annotation
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6_2" checked>
        
          
          <label class="md-nav__link" for="__nav_6_2" id="__nav_6_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Tutorial (Short)
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_6_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_6_2">
            <span class="md-nav__icon md-icon"></span>
            Tutorial (Short)
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../introduction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../reads_to_genomes/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    From Reads To Genomes
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../assigning_lineage_and_function/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Assessing Bin Quality, Lineage, and Gene Function
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Dereplication, and Coverage Inspection of MAGs
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Dereplication, and Coverage Inspection of MAGs
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#metagenomics-toolkit-execution" class="md-nav__link">
    <span class="md-ellipsis">
      Metagenomics-Toolkit Execution
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#output" class="md-nav__link">
    <span class="md-ellipsis">
      Output
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Output">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#dereplication-clusters" class="md-nav__link">
    <span class="md-ellipsis">
      Dereplication Clusters
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#read-mapping-depth-coverage-of-example-mags" class="md-nav__link">
    <span class="md-ellipsis">
      Read Mapping Depth (Coverage) of Example MAGs
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_7" >
        
          
          <label class="md-nav__link" for="__nav_7" id="__nav_7_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Developer
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_7">
            <span class="md-nav__icon md-icon"></span>
            Developer
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../developer_guidelines/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Guidelines
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../pipeline_specification/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Pipeline Specification
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../module_specification/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Module Specification
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#metagenomics-toolkit-execution" class="md-nav__link">
    <span class="md-ellipsis">
      Metagenomics-Toolkit Execution
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#output" class="md-nav__link">
    <span class="md-ellipsis">
      Output
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Output">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#dereplication-clusters" class="md-nav__link">
    <span class="md-ellipsis">
      Dereplication Clusters
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#read-mapping-depth-coverage-of-example-mags" class="md-nav__link">
    <span class="md-ellipsis">
      Read Mapping Depth (Coverage) of Example MAGs
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


  <h1>Dereplication, and Coverage Inspection of MAGs</h1>

<p>In this final part of the tutorial we <strong>aggregate</strong> the MAGs recovered from multiple samples, <strong>dereplicate</strong> them, and briefly inspect both the resulting clusters and the <strong>read-mapping depth</strong> (coverage) of some MAGs in the small example dataset.</p>
<p>Dereplication reduces a large set of MAGs to a <strong>non-redundant genome catalogue</strong> by grouping nearly identical genomes (typically using Average Nucleotide Identity, ANI) and
selecting the highest-quality MAG as the representative for each cluster. This avoids double-counting strains, simplifies downstream analyses and prevents ambiguous read mapping to many almost-identical genomes.</p>
<p>Within the Metagenomics-Toolkit, dereplication is implemented as a separate module that:</p>
<ol>
<li>takes the previous run or a table describing all MAGs (including completeness, contamination and coverage),</li>
<li>clusters similar genomes in a bottom-up fashion (using Mash and ANI distances), and</li>
<li>selects the best representative per cluster based on quality and assembly statistics.</li>
</ol>
<hr />
<h2 id="metagenomics-toolkit-execution">Metagenomics-Toolkit Execution<a class="headerlink" href="#metagenomics-toolkit-execution" title="Permanent link">&para;</a></h2>
<p>The following lines represent the part of the configuration that tells the Toolkit to run the <strong>Dereplication</strong> module on your existing results.</p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">Dereplication Configuration File Snippet</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nt">output</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;output&quot;</span>
<span class="nt">input</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;output&quot;</span>
<span class="nt">logDir</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">log</span>
<span class="nt">runid</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1</span>
<span class="nt">logLevel</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1</span>
<span class="nt">scratch</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">false</span>
<span class="nt">publishDirMode</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;copy&quot;</span>
<span class="nt">steps</span><span class="p">:</span>
<span class="w">  </span><span class="nt">dereplication</span><span class="p">:</span>
<span class="w">    </span><span class="nt">bottomUpClustering</span><span class="p">:</span>
<span class="w">      </span><span class="c1"># stricter MIMAG medium quality</span>
<span class="w">      </span><span class="nt">minimumCompleteness</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">50</span>
<span class="w">      </span><span class="nt">maximumContamination</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">5</span>
<span class="w">      </span><span class="nt">ANIBuffer</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">20</span>
<span class="w">      </span><span class="nt">mashBuffer</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">2000</span>
<span class="w">      </span><span class="nt">method</span><span class="p">:</span><span class="w"> </span><span class="s">&#39;ANI&#39;</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span>
<span class="w">        </span><span class="nt">mash_sketch</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;&quot;</span>
<span class="w">        </span><span class="nt">mash_dist</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;&quot;</span>
<span class="w">        </span><span class="c1"># cluster cutoff</span>
<span class="w">        </span><span class="nt">cluster</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">-c</span><span class="nv"> </span><span class="s">0.05</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">        </span><span class="nt">pyani</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">-m</span><span class="nv"> </span><span class="s">ANIb</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">        </span><span class="nt">representativeAniCutoff</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.95</span>
<span class="w">  </span><span class="nt">readMapping</span><span class="p">:</span>
<span class="w">    </span><span class="nt">bwa2</span><span class="p">:</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span>
<span class="w">        </span><span class="nt">bwa2_index</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;&quot;</span>
<span class="w">        </span><span class="nt">bwa2_mem</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;&quot;</span>
<span class="w">     </span><span class="c1"># This module produces two abundance tables.</span>
<span class="w">     </span><span class="c1"># One table is based on relative abundance and the second one on the trimmed mean.</span>
<span class="w">     </span><span class="c1"># Just using relative abundance makes it difficult to tell if a genome is part of a dataset.</span>
<span class="w">     </span><span class="c1"># Thats why it makes sense to set at leat a low min covered fraction parameter.</span>
<span class="w">    </span><span class="nt">coverm</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">--min-covered-fraction</span><span class="nv"> </span><span class="s">80</span><span class="nv">  </span><span class="s">--min-read-percent-identity</span><span class="nv"> </span><span class="s">95</span><span class="nv"> </span><span class="s">--min-read-aligned-percent</span><span class="nv"> </span><span class="s">95</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">    </span><span class="nt">covermONT</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">--min-covered-fraction</span><span class="nv"> </span><span class="s">80</span><span class="nv">  </span><span class="s">--min-read-aligned-percent</span><span class="nv"> </span><span class="s">95</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">    </span><span class="nt">minimap</span><span class="p">:</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span>
<span class="w">        </span><span class="nt">minimap_index</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;&quot;</span>
<span class="w">        </span><span class="nt">minimap</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;&quot;</span>
</code></pre></div></td></tr></table></div>
<p>For reference, your complete parameter file for this final step looks like this:</p>
<details class="parameter-file">
<summary>Parameter-file</summary>
<div class="highlight"><pre><span></span><code>tempdir:<span class="w"> </span><span class="s2">&quot;tmp&quot;</span>
summary:<span class="w"> </span><span class="nb">false</span>
s3SignIn:<span class="w"> </span><span class="nb">false</span>
output:<span class="w"> </span><span class="s2">&quot;output&quot;</span>
input:<span class="w"> </span><span class="s2">&quot;output&quot;</span>
logDir:<span class="w"> </span>log
runid:<span class="w"> </span><span class="m">1</span>
logLevel:<span class="w"> </span><span class="m">1</span>
scratch:<span class="w"> </span><span class="nb">false</span>
publishDirMode:<span class="w"> </span><span class="s2">&quot;copy&quot;</span>
steps:
<span class="w">  </span>dereplication:
<span class="w">    </span>bottomUpClustering:
<span class="w">      </span><span class="c1"># stricter MIMAG medium quality</span>
<span class="w">      </span>minimumCompleteness:<span class="w"> </span><span class="m">50</span>
<span class="w">      </span>maximumContamination:<span class="w"> </span><span class="m">5</span>
<span class="w">      </span>ANIBuffer:<span class="w"> </span><span class="m">20</span>
<span class="w">      </span>mashBuffer:<span class="w"> </span><span class="m">2000</span>
<span class="w">      </span>method:<span class="w"> </span><span class="s1">&#39;ANI&#39;</span>
<span class="w">      </span>additionalParams:
<span class="w">        </span>mash_sketch:<span class="w"> </span><span class="s2">&quot;&quot;</span>
<span class="w">        </span>mash_dist:<span class="w"> </span><span class="s2">&quot;&quot;</span>
<span class="w">        </span><span class="c1"># cluster cutoff</span>
<span class="w">        </span>cluster:<span class="w"> </span><span class="s2">&quot; -c 0.05 &quot;</span>
<span class="w">        </span>pyani:<span class="w"> </span><span class="s2">&quot; -m ANIb &quot;</span>
<span class="w">        </span>representativeAniCutoff:<span class="w"> </span><span class="m">0</span>.95
<span class="w">  </span>readMapping:
<span class="w">    </span>bwa2:
<span class="w">      </span>additionalParams:
<span class="w">        </span>bwa2_index:<span class="w"> </span><span class="s2">&quot;&quot;</span>
<span class="w">        </span>bwa2_mem:<span class="w"> </span><span class="s2">&quot;&quot;</span>
<span class="w">     </span><span class="c1"># This module produces two abundance tables.</span>
<span class="w">     </span><span class="c1"># One table is based on relative abundance and the second one on the trimmed mean.</span>
<span class="w">     </span><span class="c1"># Just using relative abundance makes it difficult to tell if a genome is part of a dataset.</span>
<span class="w">     </span><span class="c1"># Thats why it makes sense to set at leat a low min covered fraction parameter.</span>
<span class="w">    </span>coverm:<span class="w"> </span><span class="s2">&quot; --min-covered-fraction 80  --min-read-percent-identity 95 --min-read-aligned-percent 95 &quot;</span>
<span class="w">    </span>covermONT:<span class="w"> </span><span class="s2">&quot; --min-covered-fraction 80  --min-read-aligned-percent 95 &quot;</span>
<span class="w">    </span>minimap:
<span class="w">      </span>additionalParams:
<span class="w">        </span>minimap_index:<span class="w"> </span><span class="s2">&quot;&quot;</span>
<span class="w">        </span>minimap:<span class="w"> </span><span class="s2">&quot;&quot;</span>
resources:
<span class="w">  </span>highmemLarge:
<span class="w">    </span>cpus:<span class="w"> </span><span class="m">28</span>
<span class="w">    </span>memory:<span class="w"> </span><span class="m">230</span>
<span class="w">  </span>highmemMedium:
<span class="w">    </span>cpus:<span class="w"> </span><span class="m">14</span>
<span class="w">    </span>memory:<span class="w"> </span><span class="m">113</span>
<span class="w">  </span>large:
<span class="w">    </span>cpus:<span class="w"> </span><span class="m">28</span>
<span class="w">    </span>memory:<span class="w"> </span><span class="m">58</span>
<span class="w">  </span>medium:
<span class="w">    </span>cpus:<span class="w"> </span><span class="m">14</span>
<span class="w">    </span>memory:<span class="w"> </span><span class="m">29</span>
<span class="w">  </span>small:
<span class="w">    </span>cpus:<span class="w"> </span><span class="m">7</span>
<span class="w">    </span>memory:<span class="w"> </span><span class="m">14</span>
<span class="w">  </span>tiny:
<span class="w">    </span>cpus:<span class="w"> </span><span class="m">1</span>
<span class="w">    </span>memory:<span class="w"> </span><span class="m">1</span>
</code></pre></div>
</details>
<p>We assume that all required steps by the previous tutorials are done.</p>
<p>If you run the Toolkit again with the <code>-resume</code> flag and the above dereplication configuration, it will <strong>reuse</strong> all previous results (QC, assembly, binning, MagAttributes) and only execute the new dereplication step.</p>
<div class="admonition question">
<p class="admonition-title">Task 1</p>
<p>Change into the directory of the Toolkit session.
<div class="highlight"><pre><span></span><code><span class="nb">cd</span><span class="w"> </span>/vol/volume/sessions/metagenomics_metagenomics-tk
</code></pre></div></p>
<p>Run the dereplication module on your results using the following command:
<div class="highlight"><pre><span></span><code><span class="nv">NXF_VER</span><span class="o">=</span><span class="m">25</span>.04.2<span class="w"> </span>nextflow<span class="w"> </span>run<span class="w"> </span>metagenomics/metagenomics-tk<span class="w"> </span><span class="se">\</span>
<span class="w">      </span>-profile<span class="w"> </span>standard<span class="w"> </span><span class="se">\</span>
<span class="w">      </span>-params-file<span class="w"> </span>https://raw.githubusercontent.com/metagenomics/metagenomics-tk/refs/heads/doc/exhibition-day/tutorial/default/tutorials/tutorial2/fullPipeline_dereplication.yml<span class="w"> </span><span class="se">\</span>
<span class="w">      </span>-ansi-log<span class="w"> </span><span class="nb">false</span><span class="w"> </span><span class="se">\</span>
<span class="w">      </span>-entry<span class="w"> </span>wAggregatePipeline<span class="w"> </span><span class="se">\</span>
<span class="w">      </span>-resume<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--input<span class="w"> </span>output<span class="w"> </span><span class="se">\</span>
<span class="w">      </span>--logDir<span class="w"> </span>dereplication<span class="w"> </span>
</code></pre></div></p>
<p>This will:</p>
<ul>
<li>collect all the necessary results from your MAGs to create the summary table needed as input,</li>
<li>cluster similar genomes based on ANI, and</li>
<li>write a <code>clusters.tsv</code> file describing the dereplicated genome clusters.</li>
</ul>
</div>
<hr />
<h2 id="output">Output<a class="headerlink" href="#output" title="Permanent link">&para;</a></h2>
<h3 id="dereplication-clusters">Dereplication Clusters<a class="headerlink" href="#dereplication-clusters" title="Permanent link">&para;</a></h3>
<p>The example dereplication input table looks like this (one MAG per row):</p>
<div class="highlight"><pre><span></span><code>BIN_ID  COMPLETENESS    COVERAGE        CONTAMINATION   HETEROGENEITY   PATH    N50
SRR492065_bin.3.fa      36.29   171.259 0.39    0       metagenomics-tk/work/58/267ffbad749945cb7dfd98b23a7449/SRR492065_bin.3.fa      73239
SRR492183_bin.2.fa      100.0   142.023 0.1     0       metagenomics-tk/work/90/b5b80ef93e064e7d723eca9bd92241/SRR492183_bin.2.fa      175163
SRR492065_bin.1.fa      85.99   14.7729 18.95   0       metagenomics-tk/work/58/267ffbad749945cb7dfd98b23a7449/SRR492065_bin.1.fa      7308
SRR492065_bin.5.fa      64.75   152.837 0.54    0       metagenomics-tk/work/58/267ffbad749945cb7dfd98b23a7449/SRR492065_bin.5.fa      161674
SRR492065_bin.4.fa      96.32   70.5657 0.14    0       metagenomics-tk/work/58/267ffbad749945cb7dfd98b23a7449/SRR492065_bin.4.fa      25145
SRR492065_bin.2.fa      94.64   23.2949 0.11    0       metagenomics-tk/work/58/267ffbad749945cb7dfd98b23a7449/SRR492065_bin.2.fa      16802
SRR492183_bin.3.fa      92.84   27.7866 0.08    0       metagenomics-tk/work/90/b5b80ef93e064e7d723eca9bd92241/SRR492183_bin.3.fa      21121
SRR492183_bin.1.fa      99.96   45.5264 0.01    0       metagenomics-tk/work/90/b5b80ef93e064e7d723eca9bd92241/SRR492183_bin.1.fa      46795
</code></pre></div>
<p>The dereplication module uses these columns as follows:</p>
<ul>
<li><code>COMPLETENESS</code>, <code>CONTAMINATION</code>  to filter MAGs before clustering.</li>
<li><code>COVERAGE</code>, <code>N50</code>, <code>HETEROGENEITY</code>  to decide which genome becomes the <strong>representative</strong> of each cluster.</li>
</ul>
<p>After running Task 1, dereplication produces a <code>clusters.tsv</code> file with one row per MAG and three columns:</p>
<ul>
<li><code>CLUSTER</code>  cluster ID (all genomes in the same cluster are near-identical),</li>
<li><code>GENOME</code>  path or identifier of the genome,</li>
<li><code>REPRESENTATIVE</code>  <code>1</code> if the genome is the chosen representative, <code>0</code> otherwise.</li>
</ul>
<p>A typical file will look like this:</p>
<div class="highlight"><pre><span></span><code>CLUSTER  GENOME                                 REPRESENTATIVE
1        bin.1.fa                1
1        bin.2.fa                0
2        bin.8.fasta             1
2        bin.9.fasta             0
3        bin.10.fasta            0
3        bin.32.fa               1
</code></pre></div>
<p><em>(The exact clustering in your run may differ; this is just an illustrative example.)</em></p>
<div class="admonition question">
<p class="admonition-title">Task 2</p>
<p>Locate the dereplication clusters file and inspect how many clusters were formed and which genomes were chosen as representatives.
Hint: There is a new directory in your output folder.</p>
<details class="solution">
<summary>Solution</summary>
<p>Change into the Toolkit session directory (if you are not already there):</p>
<div class="highlight"><pre><span></span><code><span class="nb">cd</span><span class="w"> </span>/vol/volume/sessions/metagenomics_metagenomics-tk
</code></pre></div>
<p>First, locate the new directory in your output folder:</p>
<p><div class="highlight"><pre><span></span><code>ls<span class="w"> </span>-la<span class="w"> </span>output/
</code></pre></div>
There is a new AGGREGATED directory, lets look at our <code>clusters.tsv</code> in the similarly named directory.</p>
<div class="highlight"><pre><span></span><code>cat<span class="w"> </span>output/AGGREGATED/1/dereplication/0.1.1/bottomUpClustering/clusters/clusters.tsv
</code></pre></div>
<p>It should look similar to this:</p>
<div class="highlight"><pre><span></span><code>CLUSTER GENOME  REPRESENTATIVE
1       SRR492065_bin.2.fa      0.0
3       SRR492065_bin.4.fa      1.0
2       SRR492065_bin.5.fa      0.0
2       SRR492183_bin.1.fa      1.0
4       SRR492183_bin.2.fa      1.0
1       SRR492183_bin.3.fa      1.0
</code></pre></div>
<p>This shows you:
* how many <strong>clusters</strong> exist (number of unique values in the <code>CLUSTER</code> column), and<br />
* which MAG(s) per cluster have <code>REPRESENTATIVE=1</code>. Two MAGs where redundant and are now part of clusters represented only by its best MAG.</p>
</details>
</div>
<hr />
<h3 id="read-mapping-depth-coverage-of-example-mags">Read Mapping Depth (Coverage) of Example MAGs<a class="headerlink" href="#read-mapping-depth-coverage-of-example-mags" title="Permanent link">&para;</a></h3>
<p>Coverage values are derived from mapping reads from all input samples back to each representative MAG.</p>
<div class="admonition question">
<p class="admonition-title">Task 3</p>
<p>Let's have a look to examine the coverage (read mapping depth) of our representative MAGs.
Explore the new directory, find the <code>readMapping</code> folder and look for the relative abundance coverage of our representatives with all SRR492065 reads. </p>
<details class="solution">
<summary>Solution</summary>
<p>Change again into the Toolkit session directory:</p>
<div class="highlight"><pre><span></span><code><span class="nb">cd</span><span class="w"> </span>/vol/volume/sessions/metagenomics_metagenomics-tk
</code></pre></div>
<p>Explore the AGGREGATED directory:</p>
<div class="highlight"><pre><span></span><code>ls<span class="w"> </span>-la<span class="w"> </span>output/AGGREGATED/1/
</code></pre></div>
<p>There is a <code>readMapping</code> directory.
If we explore further:</p>
<p><div class="highlight"><pre><span></span><code>ls<span class="w"> </span>-la<span class="w"> </span>output/AGGREGATED/1/readMapping/0.1.0/
</code></pre></div>
Also a <code>genomeCoverage</code> one.
If we look at this, we see something like:</p>
<p><div class="highlight"><pre><span></span><code>-rw-r--r-- 1 ubuntu ubuntu  111 Dec  9 18:50 SRR492065_count.tsv
-rw-r--r-- 1 ubuntu ubuntu  120 Dec  9 18:50 SRR492065_mean.tsv
-rw-r--r-- 1 ubuntu ubuntu  138 Dec  9 18:50 SRR492065_relative_abundance.tsv
-rw-r--r-- 1 ubuntu ubuntu  120 Dec  9 18:50 SRR492065_rpkm.tsv
-rw-r--r-- 1 ubuntu ubuntu  116 Dec  9 18:50 SRR492065_tpm.tsv
-rw-r--r-- 1 ubuntu ubuntu  118 Dec  9 18:50 SRR492065_trimmed_mean.tsv
-rw-r--r-- 1 ubuntu ubuntu  110 Dec  9 18:50 SRR492183_count.tsv
-rw-r--r-- 1 ubuntu ubuntu  113 Dec  9 18:50 SRR492183_mean.tsv
-rw-r--r-- 1 ubuntu ubuntu  130 Dec  9 18:50 SRR492183_relative_abundance.tsv
-rw-r--r-- 1 ubuntu ubuntu  110 Dec  9 18:50 SRR492183_rpkm.tsv
-rw-r--r-- 1 ubuntu ubuntu  112 Dec  9 18:50 SRR492183_tpm.tsv
-rw-r--r-- 1 ubuntu ubuntu  113 Dec  9 18:50 SRR492183_trimmed_mean.tsv
drwxr-xr-x 2 ubuntu ubuntu 4096 Dec  9 18:50 logs
</code></pre></div>
In <code>SRR492065_relative_abundance.tsv</code> we find the right abundances.</p>
<div class="highlight"><pre><span></span><code>cat<span class="w"> </span>output/AGGREGATED/1/readMapping/0.1.0/genomeCoverage/SRR492065_relative_abundance.tsv
</code></pre></div>
<div class="highlight"><pre><span></span><code>Genome  SRR492065
unmapped        21.920425
SRR492065_bin.4 22.608734
SRR492183_bin.1 45.45554
SRR492183_bin.2 2.6764407
SRR492183_bin.3 7.338865
</code></pre></div>
</details>
</div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../../..", "features": ["navigation.tabs", "navigation.tabs.sticky", "content.code.copy"], "search": "../../../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"provider": "mike"}}</script>
    
    
      <script src="../../../assets/javascripts/bundle.bd41221c.min.js"></script>
      
    
  </body>
</html>