### Exploratory Metagenome Browser (EMGB)

The output generated by the Metagenomics-Toolkit can be imported into [EMGB](https://gitab.ub.uni-bielefeld.de/cmg/emgb/emgb-server).
EMGB allows you to easily explore your metagenomic samples in terms of MAGs, genes and their function.

Your configuration must include at least the following analysis steps:

* Assembly

* Binning

* Gene prediction and annotation with Prokka

* Gene annotation with MMseqs

* Gene taxonomy prediction with MMseqs

We offer a script `bin/emgb.sh` that allows you to export a metagenomics toolkit output folder to emgb compatible json files.

Example Call:

```
bash emgb.sh --output=output/test1 --runid=1 --binsdir=output/test1/1/binning/0.5.0/metabat --blastdb=bacmet20_predicted --name=test1
```

You can get a help page for the necessary arguments by running `emgb.sh --help`.


