# Plasmids

The plasmid module is able to identify contigs as plasmids and also to assemble plasmids from the samples fastq data. The module is executed in two
parts. In the first part contigs of a metagenome assembler are scanned for plasmids. In the second part a plasmid assembler is used to assemble
circular plasmids out of raw reads. All plasmid detection tools are executed on the circular assembly result and on the contigs of the metagenome assembler.
Just the filtered sequences are used for downstream analysis. 

The identification of plasmids is based on the combined result of tools which have a `filter` property assigned. Results of all tools that
have the `filter` property set to true are combined either by a logical `OR` or by a logical `AND`. 

Example for the `OR` and `AND` operations: 
Let's assume that we have three plasmid detection tools (t1, t2, t3) that have four contigs (c1, c2, c3, c4) as input. Let's further assume that c1 and c2 are detected by all tools as contigs and
c3 and c4 are only detected by t1 and t2. By using an `AND` only c1 and c2 are finally reported by the module as plasmids. By using an `OR` all contigs would be annotated as plasmids. 

It is also possible to simply run a tool without
using its result as filter by setting `filter` to `false`. If a tool should not be executed then the tool section should be removed.
Only the detected plasmids will be used for downstream analysis.


For running a plasmid assembly we suggest running the full pipeline mode with the enabled plasmids module. See input example configuration files.

## Input

=== "Command"

    ```
    -entry wPlasmidsPath -params-file example_params/plasmids.yml
    ```

=== "Configuration file for full pipeline mode with plasmids detections"

    ```YAML
    ---8<--- "../example_params/fullPipeline_fraction/fullPipeline_fraction_plasmid.yml"
    ```

=== "Configuration file for plasmids module only"

    ```YAML
    ---8<--- "../example_params/plasmid.yml"
    ```

=== "TSV Table"

    ```TSV
    ---8<--- "../test_data/plasmid/input_contigs.tsv"
    ```

=== "Graph"

    ```mermaid
    flowchart TD
        A[Metaspades or Megahit] --> C[Assembly graph] 
        subgraph circ [Circular Plasmids]
            C --> D[SCAPP]
            subgraph fc [User selected combination of tools for filtering contigs using logical OR or AND]
            direction LR
                pa[Platon]
                vi[ViralVerify]
                mo[MobTyper]
                plc[PlasClass]
            end
            D --> fc
            D --> ca[Contig Abundance]
            D --> ga[Gene Abundance]
            fc --> ap[Filtered Assembled Plasmids]
        end
        subgraph luc [Linear or Circular Plasmids]
        gac[Gene Abundance] 
        c[Contigs]
            subgraph fl [User selected combination of tools for filtering contigs using logical OR or AND]
            direction LR
                p[Platon]
                v[ViralVerify]
                m[MobTyper]
                pl[PlasClass]
            end
        d[Detected Plasmids]
        c --> cac[Contig Abundance]
        end
        A --> c --> fl
        fl --> d
    
        subgraph an [Annotation Module]
            vf[VFDB]
            ba[BacMet]
            pl[PLSDB]
            rgi[RGI]
            ke[KEGG]
            o[...]
        end
        luc --> an
        circ --> an
    ```

### Databases

The plasmid module needs the following compressed database file formats: 

* ViralVerifyPlasmid: Recent pfam-A database in .gz format.

* MobTyper: Database was generated by gzipping the output of mob_init.

* Platon: The tar gzipped database for running platon can be fetched from the [Platon](https://github.com/oschwengers/platon) github page.

* PLSDB: Database is available via this link:
         [https://ccb-microbe.cs.uni-saarland.de/plsdb/plasmids/download/plasmids_meta.tar.bz2](https://ccb-microbe.cs.uni-saarland.de/plsdb/plasmids/download/plasmids_meta.tar.bz2).
         All files except *.tsv and *.msh were deleted from the compressed package.

See [database section](../pipeline_configuration.md#database-input-configuration) for possible download strategies.

#### PLSDB

PLSDB needs a plasmid database as input. See [database section](../pipeline_configuration.md#database-input-configuration) for possible download strategies.
The compressed database must be a tar.bz2 file. 

## Output

### SCAPP

SCAPP detects plasmid sequences out of the samples assembly graph.
It reports sequences as gzipped fasta files (`*_plasmids.fasta.gz`). A basic statistic (`*_plasmids_stats.tsv`) per plasmid and a summary statistic (`*_plasmids_summary_stats.tsv`) over all
plasmids is also generated. Coverm coverage metrics are generated for all plasmids. Gene coverage values are generated as part of the annotation module output.

### PlasClass

PlasClass is able to identify plasmids by using a statistical model that was build using kmer frequencies.
It reports gzipped fata files and their probabilities (`*_plasclass.tsv`).

### MobTyper and Platon

MobTyper and Platon are using both replicon typing for plasmid detection. (`*_mobtyper_results.tsv`, `*_platon.tsv`)

### ViralVerifyPlasmid

ViralVerfiy is applying a Naive Bayes classifier (`*_viralverifyplasmid.tsv`).

### PLSDB

PLSDB includes a curated set of plasmid sequences that were extracted from databases like refseq.
The metadata of found sequences are reported in `*.tsv` and the metadata of the filtered sequences in `*_kmerThreshold_X.tsv`.

