
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../introduction/">
      
      
        <link rel="next" href="../assigning_lineage_and_function/">
      
      
      <link rel="icon" href="../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.5.14">
    
    
      
        <title>From Reads To Genomes - Metagenomics-Toolkit</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.10ba22f1.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="blue-grey" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#preparation" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      <div data-md-color-scheme="default" data-md-component="outdated" hidden>
        
          <aside class="md-banner md-banner--warning">
            <div class="md-banner__inner md-grid md-typeset">
              
  You're not viewing the latest version.
    <a href="../../../..">
	        <strong>Click here to go to latest.</strong>
    </a>

            </div>
            <script>var el=document.querySelector("[data-md-component=outdated]"),outdated=__md_get("__outdated",sessionStorage);!0===outdated&&el&&(el.hidden=!1)</script>
          </aside>
        
      </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="Metagenomics-Toolkit" class="md-header__button md-logo" aria-label="Metagenomics-Toolkit" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Metagenomics-Toolkit
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              From Reads To Genomes
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="blue-grey" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="blue-grey" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/metagenomics/metagenomics-tk" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    Metagenomics-Tk
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../.." class="md-tabs__link">
        
  
    
  
  Introduction

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../../quickstart/" class="md-tabs__link">
        
  
    
  
  Quickstart

      </a>
    </li>
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../../overview/" class="md-tabs__link">
          
  
  Getting Started

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../../configuration/" class="md-tabs__link">
          
  
  Configuration

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../../modules/introduction/" class="md-tabs__link">
          
  
  Modules

        </a>
      </li>
    
  

      
        
  
  
    
  
  
    
    
      
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../../tutorial1/introduction/" class="md-tabs__link">
          
  
  Tutorials

        </a>
      </li>
    
  

    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../../developer_guidelines/" class="md-tabs__link">
          
  
  Developer

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="Metagenomics-Toolkit" class="md-nav__button md-logo" aria-label="Metagenomics-Toolkit" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    Metagenomics-Toolkit
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/metagenomics/metagenomics-tk" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    Metagenomics-Tk
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../quickstart/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Quickstart
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Getting Started
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Getting Started
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../overview/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../full_pipeline/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Full Pipeline
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../aggregation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Aggregation
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Configuration
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Configuration
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../configuration/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Global Pipeline Configuration
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../pipeline_input/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Pipeline Input Configuration
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../database/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Database Configuration
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Modules
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Modules
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/introduction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/qualityControl/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Quality Control
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/annotation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Annotation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/assembly/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Assembly
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/cooccurrence/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Cooccurrence
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/dereplication/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Dereplication
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/fragment_recruitment/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Run Fragment Recruitment
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/magAttributes/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    MagAttributes
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/plasmids/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Plasmids
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/metabolomics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Metabolomics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/readMapping/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Read Mapping
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../modules/export/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Exploratory Metagenome Browser (EMGB)
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
      
        
        
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6" checked>
        
          
          <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Tutorials
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_6">
            <span class="md-nav__icon md-icon"></span>
            Tutorials
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6_1" >
        
          
          <label class="md-nav__link" for="__nav_6_1" id="__nav_6_1_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Tutorial (Basics)
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_6_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6_1">
            <span class="md-nav__icon md-icon"></span>
            Tutorial (Basics)
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tutorial1/introduction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tutorial1/quality_control/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Quality Control
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tutorial1/assembly/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Assembly
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tutorial1/assembly_evaluation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Assembly Evaluation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tutorial1/binning/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Binning
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tutorial1/bin_quality/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Assessing Bin Quality
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tutorial1/classification/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Classification
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tutorial1/annotation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Annotation
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6_2" checked>
        
          
          <label class="md-nav__link" for="__nav_6_2" id="__nav_6_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Tutorial (Short)
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_6_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_6_2">
            <span class="md-nav__icon md-icon"></span>
            Tutorial (Short)
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../introduction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    From Reads To Genomes
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    From Reads To Genomes
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#preparation" class="md-nav__link">
    <span class="md-ellipsis">
      Preparation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#metagenomics-toolkit-execution" class="md-nav__link">
    <span class="md-ellipsis">
      Metagenomics-Toolkit Execution
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#output" class="md-nav__link">
    <span class="md-ellipsis">
      Output
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Output">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#quality-control" class="md-nav__link">
    <span class="md-ellipsis">
      Quality Control
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Quality Control">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#fastp" class="md-nav__link">
    <span class="md-ellipsis">
      Fastp
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#host-dna-removal" class="md-nav__link">
    <span class="md-ellipsis">
      Host DNA Removal
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#assembly-results" class="md-nav__link">
    <span class="md-ellipsis">
      Assembly Results
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#binning-results" class="md-nav__link">
    <span class="md-ellipsis">
      Binning Results
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Binning Results">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#metabat" class="md-nav__link">
    <span class="md-ellipsis">
      MetaBAT
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../assigning_lineage_and_function/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Assessing Bin Quality, Lineage, and Gene Function
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../dereplication/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Dereplication, and Coverage Inspection of MAGs
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_7" >
        
          
          <label class="md-nav__link" for="__nav_7" id="__nav_7_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Developer
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_7">
            <span class="md-nav__icon md-icon"></span>
            Developer
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../developer_guidelines/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Guidelines
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../pipeline_specification/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Pipeline Specification
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../module_specification/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Module Specification
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#preparation" class="md-nav__link">
    <span class="md-ellipsis">
      Preparation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#metagenomics-toolkit-execution" class="md-nav__link">
    <span class="md-ellipsis">
      Metagenomics-Toolkit Execution
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#output" class="md-nav__link">
    <span class="md-ellipsis">
      Output
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Output">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#quality-control" class="md-nav__link">
    <span class="md-ellipsis">
      Quality Control
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Quality Control">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#fastp" class="md-nav__link">
    <span class="md-ellipsis">
      Fastp
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#host-dna-removal" class="md-nav__link">
    <span class="md-ellipsis">
      Host DNA Removal
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#assembly-results" class="md-nav__link">
    <span class="md-ellipsis">
      Assembly Results
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#binning-results" class="md-nav__link">
    <span class="md-ellipsis">
      Binning Results
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Binning Results">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#metabat" class="md-nav__link">
    <span class="md-ellipsis">
      MetaBAT
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


  <h1>From Reads To Genomes</h1>

<p>The first step is to run a thorough quality‑control pipeline that trims adapters and low‑quality ends, discards reads that become too short, and removes any host DNA contamination. 
A Nonpareil analysis is then executed to estimate the sequencing depth needed to capture the full microbial diversity.</p>
<p>With the cleaned reads in hand, the Toolkit proceeds to assembly. The MEGAHIT assembler, integrated into the Toolkit, is invoked to build contigs representing longer contiguous sequences in the sample.
By examining each contig’s k‑mer composition and its coverage, the binning tools cluster contigs that share similar signatures.
These clusters, or “Bins,” correspond to draft genomes reconstructed from the community, enabling researchers to recover and study individual organisms without prior knowledge of the sample’s composition.</p>
<p>In short, the guide describes how the Toolkit orchestrates QC, assembly and binning to turn raw metagenomic reads into biologically meaningful genome bins.</p>
<p>We will inspect two SRA samples (<a href="https://trace.ncbi.nlm.nih.gov/Traces/?view=run_browser&amp;acc=SRR492065&amp;display=metadata" target="_blank">SRR492065</a> and <a href="https://trace.ncbi.nlm.nih.gov/Traces/?view=run_browser&amp;acc=SRR492183&amp;display=metadata" target="_blank">SRR492183</a>)
that belong to the same study "Preborn infant gut metagenome".</p>
<h2 id="preparation">Preparation<a class="headerlink" href="#preparation" title="Permanent link">&para;</a></h2>
<p>Due to the configuration of your VMs, you need to create a small config file with the following command, to make sure we can resume workflow runs.</p>
<div class="admonition question">
<p class="admonition-title">Task 1</p>
<p>Change into the directory of the Toolkit session.
<div class="highlight"><pre><span></span><code><span class="nb">cd</span><span class="w"> </span>/vol/volume/sessions/metagenomics_metagenomics-tk<span class="w"> </span>
</code></pre></div></p>
<p>Copy the following command to create the config file.</p>
<div class="highlight"><pre><span></span><code><span class="nb">echo</span><span class="w"> </span><span class="s1">&#39;cleanup = false&#39;</span><span class="w"> </span>&gt;<span class="w"> </span>nextflow.config
</code></pre></div>
</div>
<h2 id="metagenomics-toolkit-execution">Metagenomics-Toolkit Execution<a class="headerlink" href="#metagenomics-toolkit-execution" title="Permanent link">&para;</a></h2>
<p>The Metagenomics-Toolkit offers multiple tools for each of the aforementioned processing steps.
You will now execute the following Toolkit configuration:</p>
<div class="highlight"><table class="highlighttable"><tr><th colspan="2" class="filename"><span class="filename">QC,Assembly,Binning Configuration File Snippet 1</span></th></tr><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nt">tempdir</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;tmp&quot;</span>
<span class="nt">summary</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">false</span>
<span class="nt">s3SignIn</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">false</span>
<span class="nt">input</span><span class="p">:</span>
<span class="w">  </span><span class="nt">paired</span><span class="p">:</span>
<span class="w">    </span><span class="nt">path</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;test_data/tutorials/tutorial2/samples.tsv&quot;</span>
<span class="w">    </span><span class="nt">watch</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">false</span>
<span class="nt">output</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">output</span>
<span class="nt">logDir</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">log</span>
<span class="nt">runid</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1</span>
<span class="nt">databases</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;/vol/volume/sessions/metagenomics_metagenomics-tk/databases/&quot;</span>
<span class="nt">publishDirMode</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;copy&quot;</span>
<span class="nt">logLevel</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1</span>
<span class="nt">scratch</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">false</span>
<span class="nt">steps</span><span class="p">:</span>
<span class="w">  </span><span class="nt">qc</span><span class="p">:</span>
<span class="w">    </span><span class="nt">fastp</span><span class="p">:</span>
<span class="w">       </span><span class="c1"># For PE data, the adapter sequence auto-detection is disabled by default since the adapters can be trimmed by overlap analysis. However, you can specify --detect_adapter_for_pe to enable it.</span>
<span class="w">       </span><span class="c1"># For PE data, fastp will run a little slower if you specify the sequence adapters or enable adapter auto-detection, but usually result in a slightly cleaner output, since the overlap analysis may fail due to sequencing errors or adapter dimers.</span>
<span class="w">       </span><span class="c1"># -q, --qualified_quality_phred       the quality value that a base is qualified. Default 15 means phred quality &gt;=Q15 is qualified.</span>
<span class="w">       </span><span class="c1"># --cut_front move a sliding window from front (5&#39;) to tail, drop the bases in the window if its mean quality is below cut_mean_quality, stop otherwise.</span>
<span class="w">       </span><span class="c1"># --length_required  reads shorter than length_required will be discarded, default is 15. (int [=15])</span>
<span class="w">       </span><span class="c1"># PE data, the front/tail trimming settings are given with -f, --trim_front1 and -t, --trim_tail1</span>
<span class="w">       </span><span class="nt">additionalParams</span><span class="p">:</span>
<span class="w">         </span><span class="nt">fastp</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">--detect_adapter_for_pe</span><span class="nv"> </span><span class="s">-q</span><span class="nv"> </span><span class="s">20</span><span class="nv"> </span><span class="s">--cut_front</span><span class="nv"> </span><span class="s">--trim_front1</span><span class="nv"> </span><span class="s">3</span><span class="nv"> </span><span class="s">--cut_tail</span><span class="nv"> </span><span class="s">--trim_tail1</span><span class="nv"> </span><span class="s">3</span><span class="nv"> </span><span class="s">--cut_mean_quality</span><span class="nv"> </span><span class="s">10</span><span class="nv"> </span><span class="s">--length_required</span><span class="nv"> </span><span class="s">50</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">         </span><span class="nt">reportOnly</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">false</span><span class="w"> </span>
<span class="w">       </span><span class="nt">timeLimit</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;AUTO&quot;</span>
<span class="w">    </span><span class="nt">nonpareil</span><span class="p">:</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">-v</span><span class="nv"> </span><span class="s">10</span><span class="nv"> </span><span class="s">-r</span><span class="nv"> </span><span class="s">1234</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">    </span><span class="nt">kmc</span><span class="p">:</span>
<span class="w">      </span><span class="nt">timeLimit</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;AUTO&quot;</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span>
<span class="w">        </span><span class="c1"># Computes k-mer distribution based on k-mer length 13 and 21</span>
<span class="w">        </span><span class="c1">#  -sm - use strict memory mode (memory limit from -m&lt;n&gt; switch will not be exceeded)</span>
<span class="w">        </span><span class="c1">#  -cs&lt;value&gt; - maximal value of a counter</span>
<span class="w">        </span><span class="nt">count</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">-sm</span><span class="nv"> </span><span class="s">-cs10000</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">        </span><span class="nt">histo</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">-cx50000</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">    </span><span class="nt">filterHuman</span><span class="p">:</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv">  </span><span class="s">&quot;</span>
<span class="w">      </span><span class="nt">database</span><span class="p">:</span>
<span class="w">        </span><span class="nt">download</span><span class="p">:</span>
<span class="w">          </span><span class="nt">source</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">https://openstack.cebitec.uni-bielefeld.de:8080/databases/human_filter.db.20231218v2.gz</span>
<span class="w">          </span><span class="nt">md5sum</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">cc92c0f926656565b1156d66a0db5a3c</span>
<span class="w">  </span><span class="nt">assembly</span><span class="p">:</span>
<span class="w">    </span><span class="nt">megahit</span><span class="p">:</span>
<span class="w">      </span><span class="c1"># --mem-flag 0 to use minimum memory, --mem-flag 1 (default) moderate memory and --mem-flag 2 all memory.</span>
<span class="w">      </span><span class="c1"># meta-sensitive: &#39;--min-count 1 --k-list 21,29,39,49,...,129,141&#39; </span>
<span class="w">      </span><span class="c1"># meta-large:  &#39;--k-min  27  --k-max 127 --k-step 10&#39; (large &amp; complex metagenomes, like soil)</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">--min-contig-len</span><span class="nv"> </span><span class="s">1000</span><span class="nv"> </span><span class="s">--presets</span><span class="nv"> </span><span class="s">meta-sensitive</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">      </span><span class="nt">fastg</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">true</span>
<span class="w">      </span><span class="nt">resources</span><span class="p">:</span>
<span class="w">        </span><span class="nt">RAM</span><span class="p">:</span>
<span class="w">          </span><span class="nt">mode</span><span class="p">:</span><span class="w"> </span><span class="s">&#39;PREDICT&#39;</span>
<span class="w">          </span><span class="nt">predictMinLabel</span><span class="p">:</span><span class="w"> </span><span class="s">&#39;medium&#39;</span>
<span class="w">  </span><span class="nt">binning</span><span class="p">:</span>
<span class="w">    </span><span class="nt">bwa2</span><span class="p">:</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span><span class="w"> </span>
<span class="w">        </span><span class="nt">bwa2</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">        </span><span class="c1"># samtools flags are used to filter resulting bam file</span>
<span class="w">        </span><span class="nt">samtoolsView</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">-F</span><span class="nv"> </span><span class="s">3584</span><span class="nv"> </span><span class="s">&quot;</span><span class="w"> </span>
<span class="w">    </span><span class="nt">contigsCoverage</span><span class="p">:</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">--min-covered-fraction</span><span class="nv"> </span><span class="s">0</span><span class="nv"> </span><span class="s">--min-read-percent-identity</span><span class="nv"> </span><span class="s">100</span><span class="nv"> </span><span class="s">--min-read-aligned-percent</span><span class="nv"> </span><span class="s">100</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">    </span><span class="nt">genomeCoverage</span><span class="p">:</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">--min-covered-fraction</span><span class="nv"> </span><span class="s">0</span><span class="nv"> </span><span class="s">--min-read-percent-identity</span><span class="nv"> </span><span class="s">100</span><span class="nv"> </span><span class="s">--min-read-aligned-percent</span><span class="nv"> </span><span class="s">100</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="w">    </span><span class="c1"># Primary binning tool</span>
<span class="w">    </span><span class="nt">metabat</span><span class="p">:</span>
<span class="w">      </span><span class="c1"># Set --seed positive numbers to reproduce the result exactly. Otherwise, random seed will be set each time.</span>
<span class="w">      </span><span class="nt">additionalParams</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;</span><span class="nv"> </span><span class="s">--seed</span><span class="nv"> </span><span class="s">234234</span><span class="nv">  </span><span class="s">&quot;</span>
<span class="nt">resources</span><span class="p">:</span>
<span class="w">  </span><span class="nt">highmemLarge</span><span class="p">:</span>
<span class="w">    </span><span class="nt">cpus</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">28</span>
<span class="w">    </span><span class="nt">memory</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">230</span>
<span class="w">  </span><span class="nt">highmemMedium</span><span class="p">:</span>
<span class="w">    </span><span class="nt">cpus</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">14</span>
<span class="w">    </span><span class="nt">memory</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">113</span>
<span class="w">  </span><span class="nt">large</span><span class="p">:</span>
<span class="w">    </span><span class="nt">cpus</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">28</span>
<span class="w">    </span><span class="nt">memory</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">58</span>
<span class="w">  </span><span class="nt">medium</span><span class="p">:</span>
<span class="w">    </span><span class="nt">cpus</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">14</span>
<span class="w">    </span><span class="nt">memory</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">29</span>
<span class="w">  </span><span class="nt">small</span><span class="p">:</span>
<span class="w">    </span><span class="nt">cpus</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">7</span>
<span class="w">    </span><span class="nt">memory</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">14</span>
<span class="w">  </span><span class="nt">tiny</span><span class="p">:</span>
<span class="w">    </span><span class="nt">cpus</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1</span>
<span class="w">    </span><span class="nt">memory</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1</span>
</code></pre></div></td></tr></table></div>
<div class="admonition question">
<p class="admonition-title">Task 1</p>
<p>You saw the introduction part of the tutorial regarding the structure of the config file. 
Can you tell which tools or methods are executed in the QC part?</p>
<details class="solution">
<summary>Solution</summary>
<p>Fastp, nonpareil, filterHuman, kmc</p>
</details>
</div>
<p>The following configuration runs the tools </p>
<div class="admonition question">
<p class="admonition-title">Task 2</p>
<p>Change into the directory of the Toolkit session.
<div class="highlight"><pre><span></span><code><span class="nb">cd</span><span class="w"> </span>/vol/volume/sessions/metagenomics_metagenomics-tk<span class="w"> </span>
</code></pre></div></p>
<p>Copy the following command to execute the Toolkit. The Toolkit will need about 10 to 12 minutes to complete.</p>
<div class="highlight"><pre><span></span><code><span class="nv">NXF_VER</span><span class="o">=</span><span class="m">25</span>.04.2<span class="w"> </span>nextflow<span class="w"> </span>run<span class="w"> </span>metagenomics/metagenomics-tk<span class="w"> </span><span class="se">\</span>
<span class="w">      </span>-profile<span class="w"> </span>standard<span class="w"> </span><span class="se">\</span>
<span class="w">      </span>-params-file<span class="w"> </span>https://raw.githubusercontent.com/metagenomics/metagenomics-tk/refs/heads/doc/exhibition-day/tutorial/default/tutorials/tutorial2/fullPipeline_reads_to_genomes.yml<span class="w"> </span><span class="se">\</span>
<span class="w">      </span>-ansi-log<span class="w"> </span><span class="nb">false</span><span class="w"> </span><span class="se">\</span>
<span class="w">      </span>-entry<span class="w"> </span>wFullPipeline<span class="w"> </span><span class="se">\</span>
<span class="w">      </span>-resume<span class="w"> </span><span class="se">\</span>
<span class="w">      </span>--input.paired.path<span class="w"> </span>https://raw.githubusercontent.com/metagenomics/metagenomics-tk/refs/heads/doc/exhibition-day/tutorial/test_data/tutorials/tutorial2/samples.tsv<span class="w"> </span><span class="se">\</span>
<span class="w">      </span>--logDir<span class="w"> </span>logs_reads_to_genomes<span class="w"> </span>
</code></pre></div>
</div>
<h2 id="output">Output<a class="headerlink" href="#output" title="Permanent link">&para;</a></h2>
<p>In the following we will inspect the outputs of each analysis step.</p>
<h3 id="quality-control">Quality Control<a class="headerlink" href="#quality-control" title="Permanent link">&para;</a></h3>
<h4 id="fastp">Fastp<a class="headerlink" href="#fastp" title="Permanent link">&para;</a></h4>
<p>Fastp is an efficient, versatile, open-source tool for preprocessing FASTQ files and quality control in next-generation sequencing (NGS) workflows.
For quality and adapter trimming fastp trims reads from 5’ end to 3’ end using a sliding window. </p>
<p>If the mean quality of bases inside a window drops below a specific q-score, the remainder of the read will be trimmed.
If a read gets too short during this trimming, it will be discarded.</p>
<div class="admonition info">
<p class="admonition-title">Quality Report</p>
<p>The Metagenomics-Toolkit allows you to run fastp by only reporting the quality of the data. This way you can easily inspect the data
before actually further processing it. You can do this by setting the <code>reportOnly</code> parameter to <code>true</code> in the config below. </p>
</div>
<div class="admonition question">
<p class="admonition-title">Task 3</p>
<p>Change into the directory of the Toolkit session.
<div class="highlight"><pre><span></span><code><span class="nb">cd</span><span class="w"> </span>/vol/volume/sessions/metagenomics_metagenomics-tk<span class="w"> </span>
</code></pre></div></p>
<p>You can view the fastp output in the qc output directory:
<div class="highlight"><pre><span></span><code>ls<span class="w"> </span>-1<span class="w"> </span>output/*/1/qc/*/fastp/
</code></pre></div></p>
<p>which results in the following output:</p>
<div class="highlight"><pre><span></span><code>output/SRR492065/1/qc/0.4.0/fastp:
SRR492065_fastp.json
SRR492065_fastp_summary_after.tsv
SRR492065_fastp_summary_before.tsv
SRR492065_interleaved.qc.fq.gz
SRR492065_report.html
SRR492065_unpaired.qc.fq.gz
SRR492065_unpaired_summary.tsv

output/SRR492183/1/qc/0.4.0/fastp:
SRR492183_fastp.json
SRR492183_fastp_summary_after.tsv
SRR492183_fastp_summary_before.tsv
SRR492183_interleaved.qc.fq.gz
SRR492183_report.html
SRR492183_unpaired.qc.fq.gz
SRR492183_unpaired_summary.tsv
</code></pre></div>
</div>
<p>The fastp output for your data will look like the <a href="https://s3.bi.denbi.de/cmg/mgcourses/mg2025/qc/SRR492065_report.html" target="_blank">following</a>.</p>
<div class="admonition question">
<p class="admonition-title">Task 4</p>
<p>How many reads of the wastewater dataset passed the filter?
Between which quality score ranges does the read1 FASTQ file fall before and after QC.</p>
<details class="solution">
<summary>Solution</summary>
<p>90.374006% of the reads passed the filter. The quality score for read1 ranges between 26 and 38 before filtering and after QC between 34 and 39. </p>
</details>
</div>
<h4 id="host-dna-removal">Host DNA Removal<a class="headerlink" href="#host-dna-removal" title="Permanent link">&para;</a></h4>
<p>The Toolkit provides the SRA human-scrubber tool that uses a k-mer based approach to search for human sequences against a k-mer database using human reference sequences.</p>
<div class="admonition question">
<p class="admonition-title">Task 5</p>
<p>You can find the result of the tool in the <strong>filterHuman</strong> directory:
The output of the tool are the filtered sequences, and it also includes statistics about the sequences <strong>before</strong> and <strong>after</strong> filtering.   </p>
<p>Change into the directory of the Toolkit session.
<div class="highlight"><pre><span></span><code><span class="nb">cd</span><span class="w"> </span>/vol/volume/sessions/metagenomics_metagenomics-tk<span class="w"> </span>
</code></pre></div></p>
<div class="highlight"><pre><span></span><code>ls<span class="w"> </span>output/*/1/qc/*/filterHuman/
</code></pre></div>
<p>The following command tells you how many sequences are left <strong>after</strong> filtering:</p>
<div class="highlight"><pre><span></span><code>column<span class="w"> </span>-t<span class="w">  </span>-s<span class="s1">$&#39;\t&#39;</span><span class="w"> </span>output/*/1/qc/*/filterHuman/*_interleaved_summary_*
</code></pre></div>
<p>The <strong>num_seqs</strong> column tells you the number of sequences.</p>
<p>How many reads were removed from the respective dataset after filtering?</p>
<details class="solution">
<summary>Solution</summary>
<p>3768 sequences were detected for SRR492065 and 644 for SRR492183.  </p>
</details>
</div>
<p>Please note that it is important to remove human reads before publishing read datasets, since, according to Bush et al. [^1], even a small number of reads can be used to impute phenotypes.  </p>
<hr />
<h3 id="assembly-results">Assembly Results<a class="headerlink" href="#assembly-results" title="Permanent link">&para;</a></h3>
<p>Once your raw reads have been quality-controlled, we will now perform an assembly using the Metagenomics-Toolkit.
The Toolkit supports several assemblers, including MEGAHIT, which we will run in this section.</p>
<div class="admonition question">
<p class="admonition-title">Task 6</p>
<p>Which additional parameters are used by MEGAHIT?</p>
<details class="solution">
<summary>Solution</summary>
<p>MEGAHIT with the additional parameters: minimum contig length of <code>500 bp</code> and the preset <code>meta-sensitive</code>.
<code>meta-sensitive</code> sets the following parameters: <code>--min-count 1 --k-list 21,29,39,49,...,129,141</code>, which causes 
MEGAHIT to use a longer list of k-mers. See the <a href="https://github.com/voutcn/megahit/">MEGAHIT home page</a> for more
info.</p>
</details>
</div>
<p>We will now have a first look at some assembly statistics. First of all, locate your assembly results in your <code>output</code> directory for the SRR492065 sample.</p>
<div class="admonition question">
<p class="admonition-title">Task 7</p>
<p>Where are the assembly results of MEGAHIT stored? And what files are generated per sample?</p>
<details class="solution">
<summary>Solution</summary>
<p>Change into the directory of the Toolkit session.
<div class="highlight"><pre><span></span><code>cd /vol/volume/sessions/metagenomics_metagenomics-tk
</code></pre></div>
The assembly results are stored in <code>output/data/1/assembly/1.2.1/megahit/</code>
<div class="highlight"><pre><span></span><code>ls<span class="w"> </span>-1<span class="w"> </span>output/SRR492065/1/assembly/1.2.3/megahit/
</code></pre></div>
Output:
<div class="highlight"><pre><span></span><code><span class="w">    </span>SRR492065_contigs.fa.gz<span class="w">  </span><span class="c1"># the assembled sequences as gzipped fasta</span>
<span class="w">    </span>SRR492065_contigs.fastg<span class="w">  </span><span class="c1"># the assembly graph for inspection for example with Bandage</span>
<span class="w">    </span>SRR492065_contigs_stats.tsv<span class="w">  </span><span class="c1"># some assembly statistics</span>
</code></pre></div></p>
</details>
</div>
<div class="admonition question">
<p class="admonition-title">Task 8</p>
<p>Have a look at the <code>SRR492065_contigs_stats.tsv</code> file - how large is the assembly and what is the N50?</p>
<details class="solution">
<summary>Solution</summary>
<p>Change into the directory of the Toolkit session.
<div class="highlight"><pre><span></span><code><span class="nb">cd</span><span class="w"> </span>/vol/volume/sessions/metagenomics_metagenomics-tk
</code></pre></div>
View the contig stats:
<div class="highlight"><pre><span></span><code>cat<span class="w"> </span>output/SRR492065/1/assembly/1.2.3/megahit/SRR492065_contigs_stats.tsv<span class="w"> </span><span class="p">|</span><span class="w"> </span>column<span class="w"> </span>-s<span class="s1">$&#39;\t&#39;</span><span class="w"> </span>-t
</code></pre></div>
Output: 
<div class="highlight"><pre><span></span><code>SAMPLE<span class="w">     </span>file<span class="w">                     </span>format<span class="w">  </span><span class="nb">type</span><span class="w">  </span>num_seqs<span class="w">  </span>sum_len<span class="w">   </span>min_len<span class="w">  </span>avg_len<span class="w">  </span>max_len<span class="w">  </span>Q1<span class="w">      </span>Q2<span class="w">      </span>Q3<span class="w">      </span>sum_gap<span class="w">  </span>N50<span class="w">    </span>Q20<span class="o">(</span>%<span class="o">)</span><span class="w">  </span>Q30<span class="o">(</span>%<span class="o">)</span><span class="w">  </span>GC<span class="o">(</span>%<span class="o">)</span>
SRR492065<span class="w">  </span>SRR492065_contigs.fa.gz<span class="w">  </span>FASTA<span class="w">   </span>DNA<span class="w">   </span><span class="m">1577</span><span class="w">      </span><span class="m">11206329</span><span class="w">  </span><span class="m">1000</span><span class="w">     </span><span class="m">7106</span>.1<span class="w">   </span><span class="m">245235</span><span class="w">   </span><span class="m">1554</span>.0<span class="w">  </span><span class="m">2659</span>.0<span class="w">  </span><span class="m">6398</span>.0<span class="w">  </span><span class="m">0</span><span class="w">        </span><span class="m">16801</span><span class="w">  </span><span class="m">0</span>.00<span class="w">    </span><span class="m">0</span>.00<span class="w">    </span><span class="m">40</span>.46
</code></pre></div>
In this assembly run, the assembly length is  11,206,329 bp and the N50 is 16,801 bp. </p>
</details>
</div>
<h3 id="binning-results">Binning Results<a class="headerlink" href="#binning-results" title="Permanent link">&para;</a></h3>
<p>Binning is a process in metagenomics that involves grouping DNA fragments, known as contigs,
into bins that likely originate from the same organism or genome.
This technique is essential for reconstructing genomes from mixed microbial communities,
where DNA from various organisms is intermingled, and there is often no prior knowledge of the species present.</p>
<h4 id="metabat">MetaBAT<a class="headerlink" href="#metabat" title="Permanent link">&para;</a></h4>
<p>MetaBAT is an automated metagenome binning software
which integrates empirical probabilistic distances of genome abundance
and tetranucleotide frequency. See the <a href="https://bitbucket.org/berkeleylab/metabat">MetaBAT home page</a> for more information.</p>
<p>Let's now inspect the bins created by MetaBAT:</p>
<div class="admonition question">
<p class="admonition-title">Task 9</p>
<p>How many bins did metabat generate for the dataset SRR492065? Locate the metabat results and the fasta files for each bin in the output folder.</p>
<details class="solution">
<summary>Solution</summary>
<p>Change into the directory of the Toolkit session.
<div class="highlight"><pre><span></span><code><span class="nb">cd</span><span class="w"> </span>/vol/volume/sessions/metagenomics_metagenomics-tk
</code></pre></div>
List the output of the MetaBAT tool: 
<div class="highlight"><pre><span></span><code>ls<span class="w"> </span>-1<span class="w"> </span>output/SRR492065/1/binning/*/metabat/
</code></pre></div></p>
<p>Below you can see that MetaBAT generated 5 bins. The file <code>SRR492065_notBinned.fa</code> contains contigs that could not be binned.
<div class="highlight"><pre><span></span><code>SRR492065_bin.1.fa
SRR492065_bin.2.fa
SRR492065_bin.3.fa
SRR492065_bin.4.fa
SRR492065_bin.5.fa
SRR492065_bin_contig_mapping.tsv
SRR492065_bins_stats.tsv
SRR492065_contigs_depth.tsv
SRR492065_notBinned.fa
</code></pre></div></p>
</details>
</div>
<p>The next question you might want to ask is whether you can trust these bins and which organism they represent according to a taxonomy.
Before that, let's have a look at what other information the Metagenomics-Toolkit provides as part of the binning output.</p>
<div class="admonition question">
<p class="admonition-title">Task 10</p>
<p>There is a file containing some statistics on the generated bins <code>SRR492065_bins_stats.tsv</code>. 
Find out, which bin has the highest coverage and which one has the highest N50.</p>
<details class="solution">
<summary>Solution</summary>
<p>Change into the directory of the Toolkit session.
<div class="highlight"><pre><span></span><code>cd /vol/volume/sessions/metagenomics_metagenomics-tk
</code></pre></div>
You can have a look on some bin statistics with:
<div class="highlight"><pre><span></span><code>cat<span class="w"> </span>output/SRR492065/1/binning/*/metabat/SRR492065_bins_stats.tsv<span class="w">  </span><span class="p">|</span><span class="w"> </span>column<span class="w"> </span>-s<span class="s1">$&#39;\t&#39;</span><span class="w"> </span>-t
</code></pre></div>
Output:
<div class="highlight"><pre><span></span><code>SAMPLE<span class="w">     </span>BIN_ID<span class="w">              </span>format<span class="w">  </span><span class="nb">type</span><span class="w">  </span>num_seqs<span class="w">  </span>sum_len<span class="w">  </span>min_len<span class="w">  </span>avg_len<span class="w">  </span>max_len<span class="w">  </span>Q1<span class="w">       </span>Q2<span class="w">       </span>Q3<span class="w">        </span>sum_gap<span class="w">  </span>N50<span class="w">     </span>Q20<span class="o">(</span>%<span class="o">)</span><span class="w">  </span>Q30<span class="o">(</span>%<span class="o">)</span><span class="w">  </span>GC<span class="o">(</span>%<span class="o">)</span><span class="w">  </span>COVERAGE
SRR492065<span class="w">  </span>SRR492065_bin.1.fa<span class="w">  </span>FASTA<span class="w">   </span>DNA<span class="w">   </span><span class="m">16</span><span class="w">        </span><span class="m">1536873</span><span class="w">  </span><span class="m">4900</span><span class="w">     </span><span class="m">96054</span>.6<span class="w">  </span><span class="m">245235</span><span class="w">   </span><span class="m">22104</span>.0<span class="w">  </span><span class="m">83532</span>.5<span class="w">  </span><span class="m">158838</span>.0<span class="w">  </span><span class="m">0</span><span class="w">        </span><span class="m">161674</span><span class="w">  </span><span class="m">0</span>.00<span class="w">    </span><span class="m">0</span>.00<span class="w">    </span><span class="m">36</span>.70<span class="w">  </span><span class="m">151</span>.714
SRR492065<span class="w">  </span>SRR492065_bin.2.fa<span class="w">  </span>FASTA<span class="w">   </span>DNA<span class="w">   </span><span class="m">20</span><span class="w">        </span><span class="m">1280335</span><span class="w">  </span><span class="m">3167</span><span class="w">     </span><span class="m">64016</span>.8<span class="w">  </span><span class="m">238570</span><span class="w">   </span><span class="m">23905</span>.5<span class="w">  </span><span class="m">45943</span>.5<span class="w">  </span><span class="m">65927</span>.0<span class="w">   </span><span class="m">0</span><span class="w">        </span><span class="m">185023</span><span class="w">  </span><span class="m">0</span>.00<span class="w">    </span><span class="m">0</span>.00<span class="w">    </span><span class="m">38</span>.14<span class="w">  </span><span class="m">169</span>.393
SRR492065<span class="w">  </span>SRR492065_bin.3.fa<span class="w">  </span>FASTA<span class="w">   </span>DNA<span class="w">   </span><span class="m">174</span><span class="w">       </span><span class="m">2345130</span><span class="w">  </span><span class="m">2626</span><span class="w">     </span><span class="m">13477</span>.8<span class="w">  </span><span class="m">68479</span><span class="w">    </span><span class="m">6239</span>.0<span class="w">   </span><span class="m">9839</span>.5<span class="w">   </span><span class="m">16573</span>.0<span class="w">   </span><span class="m">0</span><span class="w">        </span><span class="m">16802</span><span class="w">   </span><span class="m">0</span>.00<span class="w">    </span><span class="m">0</span>.00<span class="w">    </span><span class="m">63</span>.56<span class="w">  </span><span class="m">23</span>.2949
SRR492065<span class="w">  </span>SRR492065_bin.4.fa<span class="w">  </span>FASTA<span class="w">   </span>DNA<span class="w">   </span><span class="m">436</span><span class="w">       </span><span class="m">2720223</span><span class="w">  </span><span class="m">2505</span><span class="w">     </span><span class="m">6239</span>.0<span class="w">   </span><span class="m">36553</span><span class="w">    </span><span class="m">3386</span>.0<span class="w">   </span><span class="m">4642</span>.5<span class="w">   </span><span class="m">7408</span>.5<span class="w">    </span><span class="m">0</span><span class="w">        </span><span class="m">7308</span><span class="w">    </span><span class="m">0</span>.00<span class="w">    </span><span class="m">0</span>.00<span class="w">    </span><span class="m">32</span>.98<span class="w">  </span><span class="m">14</span>.7728
SRR492065<span class="w">  </span>SRR492065_bin.5.fa<span class="w">  </span>FASTA<span class="w">   </span>DNA<span class="w">   </span><span class="m">96</span><span class="w">        </span><span class="m">1522358</span><span class="w">  </span><span class="m">2582</span><span class="w">     </span><span class="m">15857</span>.9<span class="w">  </span><span class="m">63297</span><span class="w">    </span><span class="m">6157</span>.0<span class="w">   </span><span class="m">10569</span>.0<span class="w">  </span><span class="m">23635</span>.5<span class="w">   </span><span class="m">0</span><span class="w">        </span><span class="m">25145</span><span class="w">   </span><span class="m">0</span>.00<span class="w">    </span><span class="m">0</span>.00<span class="w">    </span><span class="m">29</span>.95<span class="w">  </span><span class="m">70</span>.5654
</code></pre></div>
In this result, Bin 2 has the highest coverage (169.393) and highest N50 (185023).</p>
</details>
</div>
<hr />
<p>➡️ <a href="../assigning_lineage_and_function/"><strong>Continue to: Assessing Bin Quality, Lineage, and Gene Function</strong></a></p>
<p>[^1:] https://pmc.ncbi.nlm.nih.gov/articles/PMC7478626/</p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../../..", "features": ["navigation.tabs", "navigation.tabs.sticky", "content.code.copy"], "search": "../../../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"provider": "mike"}}</script>
    
    
      <script src="../../../assets/javascripts/bundle.bd41221c.min.js"></script>
      
    
  </body>
</html>